{"cells":[{"cell_type":"markdown","source":["# Processing genomic data in Spark\n\n*You can import this notebook from [here](https://github.com/evodify/genomic-analyses_in_apache-spark/tree/master/vcf_filtering_tutorial).*\n\n## [Big Data: Astronomical or Genomical?](http://journals.plos.org/plosbiology/article?id=10.1371/journal.pbio.1002195)\n\n### Abstract\n\nGenomics is a Big Data science and is going to get much bigger, very soon, but it is not known whether the needs of genomics will exceed other Big Data domains. Projecting to the year 2025, we compared genomics with three other major generators of Big Data: astronomy, YouTube, and Twitter. Our estimates show that genomics is a “four-headed beast”—it is either on par with or the most demanding of the domains analyzed here in terms of data acquisition, storage, distribution, and analysis. We discuss aspects of new technologies that will need to be developed to rise up and meet the computational challenges that genomics poses for the near future. Now is the time for concerted, community-wide planning for the “genomical” challenges of the next decade.\n\n<img src=\"http://journals.plos.org/plosbiology/article/figure/image?size=large&id=10.1371/journal.pbio.1002195.g001\" width=\"800\">\n\nThe plot shows the growth of DNA sequencing both in the total number of human genomes sequenced (left axis) as well as the worldwide annual sequencing capacity (right axis: Tera-basepairs (Tbp), Peta-basepairs (Pbp), Exa-basepairs (Ebp), Zetta-basepairs (Zbps)).\n\n\n<img src=\"https://upload.wikimedia.org/wikipedia/commons/thumb/e/e7/Historic_cost_of_sequencing_a_human_genome.svg/800px-Historic_cost_of_sequencing_a_human_genome.svg.png\" width=\"800\">"],"metadata":{}},{"cell_type":"markdown","source":["## DNA sequencing\n\n<img src=\"https://www.yourgenome.org/sites/default/files/illustrations/process/physical_mapping_STS_yourgenome.png\" width=\"800\">"],"metadata":{}},{"cell_type":"markdown","source":["## Alligning millions of small DNA sequences (reads) to a reference genome \n\n<img src=\"https://github.com/evodify/genomic-analyses_in_apache-spark/raw/master/vcf_filtering_tutorial/genome_read_mapping.png\" width=\"1000\">"],"metadata":{}},{"cell_type":"markdown","source":["# VCF\n\n<img src=\"https://hail.is/docs/stable/_images/hail-vds-rep.png\" width=\"600\">"],"metadata":{}},{"cell_type":"code","source":["%scala\n\n// This allows easy embedding of publicly available information into any other notebook\n// when viewing in git-book just ignore this block - you may have to manually chase the URL in frameIt(\"URL\").\n// Example usage:\n// displayHTML(frameIt(\"https://en.wikipedia.org/wiki/Latent_Dirichlet_allocation#Topics_in_LDA\",250))\ndef frameIt( u:String, h:Int ) : String = {\n      \"\"\"<iframe \n src=\"\"\"\"+ u+\"\"\"\"\n width=\"95%\" height=\"\"\"\" + h + \"\"\"\"\n sandbox>\n  <p>\n    <a href=\"http://spark.apache.org/docs/latest/index.html\">\n      Fallback link for browsers that, unlikely, don't support frames\n    </a>\n  </p>\n</iframe>\"\"\"\n   }\ndisplayHTML(frameIt(\"https://en.wikipedia.org/wiki/Variant_Call_Format\",500))\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<iframe \n src=\"https://en.wikipedia.org/wiki/Variant_Call_Format\"\n width=\"95%\" height=\"500\"\n sandbox>\n  <p>\n    <a href=\"http://spark.apache.org/docs/latest/index.html\">\n      Fallback link for browsers that, unlikely, don't support frames\n    </a>\n  </p>\n</iframe>"]}}],"execution_count":5},{"cell_type":"markdown","source":["# Hail\n\n[Hail](https://hail.is) is an open-source, scalable framework for exploring and analyzing genomic data. Its functionality is exposed through **Python** and backed by distributed algorithms built on top of **Apache Spark** to efficiently analyze gigabyte-scale data on a laptop or terabyte-scale data on a cluster, without the need to manually chop up data or manage job failures. Users can script pipelines or explore data interactively through **Jupyter notebooks** that flow between Hail with methods for genomics, *PySpark* with scalable *SQL* and *machine learning algorithms*, and *pandas* with *scikit-learn* and *Matplotlib* for results that fit on one machine. Hail also provides a flexible domain language to express complex quality control and analysis pipelines with concise, readable code.\n\n#### Scaling Genetic Data Analysis with Apache Spark\n[![Scaling Genetic Data Analysis with Apache Spark](http://img.youtube.com/vi/pyeQusIN5Ao/0.jpg)](https://www.youtube.com/embed/pyeQusIN5Ao)"],"metadata":{}},{"cell_type":"markdown","source":["## This Notebook is based on the tutorial [Analyzing 1000 Genomes with Spark and Hail](https://docs.databricks.com/spark/latest/training/1000-genomes.html)"],"metadata":{}},{"cell_type":"markdown","source":["## Cluster setup\n\nFirst download Hail's Python and Java libraries to your computer:\n\nhttps://storage.googleapis.com/hail-common/hail-tutorial-databricks.jar\n\nhttps://storage.googleapis.com/hail-common/hail-devel-py2.7-databricks.egg\n\nThen on the Databricks interface, navigate to `Workspace > Users > Username` and select `Import` from the Username drop-down menu. At the bottom of `Import Notebooks` window, click the link in `(To import a library, such as a jar or egg,`_`click here`_`)`.  Upload both the .jar and .egg files using this interface, using any names you like. Make sure that the option `Attach automatically to all clusters` is checked in the success dialog.\n\nNext click the `Clusters` icon on the left sidebar and then `+Create Cluster`. For `Apache Spark Version`, select `Spark 2.0 (Auto-updating, Scala 2.11)`. Note that Hail won't work with Scala 2.10! In the Databricks cluster creation dialog, click `Show advanced settings` at bottom and then on the `Spark` tab, and paste the text below into the `Spark config` box.\n\n```\nspark.hadoop.io.compression.codecs org.apache.hadoop.io.compress.DefaultCodec,is.hail.io.compress.BGzipCodec,org.apache.hadoop.io.compress.GzipCodec\nspark.sql.files.openCostInBytes 1099511627776\nspark.sql.files.maxPartitionBytes 1099511627776\nspark.hadoop.mapreduce.input.fileinputformat.split.minsize 1099511627776\nspark.hadoop.parquet.block.size 1099511627776```\nStart the cluster and attach this notebook to it by clicking on your cluster name in menu `Detached` at the top left of this workbook. Now you're ready to Hail!"],"metadata":{}},{"cell_type":"code","source":["from hail import *\nhc = HailContext(sc)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":9},{"cell_type":"code","source":["#"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":10},{"cell_type":"code","source":["\n# these are the methods possible for hc ...\n%python\nfor i in dir(hc):  \n  if not i.startswith(\"_\"):\n    print(i)\n    "],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">ERROR: Line magic function &#96;%python&#96; not found (But cell magic &#96;%%python&#96; exists, did you mean that instead?).\nbalding_nichols_model\ndataframe_to_keytable\ngrep\nimport_annotations_table\nimport_bgen\nimport_gen\nimport_keytable\nimport_plink\nimport_vcf\nindex_bgen\nread\nsc\nstop\nwrite_partitioning\n</div>"]}}],"execution_count":11},{"cell_type":"markdown","source":["Let's import some Python libraries for use throughout the tutorial."],"metadata":{}},{"cell_type":"code","source":["import pandas as pd\nimport matplotlib.pyplot as plt\nimport matplotlib.patches as mpatches\nimport numpy as np\nfrom math import log, isnan\nimport seaborn"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":13},{"cell_type":"markdown","source":["## Import data\n\nWe must first import variant data into Hail's internal format of Variant Dataset (VDS). We use the [import_vcf](https://hail.is/hail/hail.HailContext.html#hail.HailContext.import_vcf) method on [HailContext](https://hail.is/hail/hail.HailContext.html) to load a VCF file into Hail.\n\nIt is recommended to load a block-compressed VCF (`.vcf.bgz`) which enables Hail to read the file in parallel. Reading files that have not been block-compressed (`.vcf`, `.vcf.gz`) is _significantly_ slower and should be avoided (though often `.vcf.gz` files are in fact block-compressed, in which case renaming to `.vcf.bgz` solves the problem).\n\nUnfortunately, I was not able to load `.bgz` compressed file. It worked for the existing human dataset, which is bgz compressed, but not for my data (See below).\n\n### Cbp data\nAs a training data set, I use a subset of 1% from my unpublished genomic data on [*Capsella bursa-pastoris*](https://en.wikipedia.org/wiki/Capsella_bursa-pastoris) (hereafter Cbp).\n\n### Human data\n\nYou can use publicly avaliable human data by uncommenting the three line below. Jump to Cmd 19."],"metadata":{}},{"cell_type":"markdown","source":["## Download the data"],"metadata":{}},{"cell_type":"code","source":["%sh\nwget https://github.com/evodify/genomic-analyses_in_apache-spark/raw/master/vcf_filtering_tutorial/Cbp31_SNPs_test0.01.vcf.gz # download the test VCF file\nwget https://github.com/evodify/genomic-analyses_in_apache-spark/raw/master/vcf_filtering_tutorial/Cbp31_annot.csv # download the test annotation file\ngunzip Cbp31_SNPs_test0.01.vcf.gz # uncomress the gzip file.\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">--2019-06-24 20:47:08--  https://github.com/evodify/genomic-analyses_in_apache-spark/raw/master/vcf_filtering_tutorial/Cbp31_SNPs_test0.01.vcf.gz\nResolving github.com (github.com)... 192.30.255.112\nConnecting to github.com (github.com)|192.30.255.112|:443... connected.\nHTTP request sent, awaiting response... 302 Found\nLocation: https://raw.githubusercontent.com/evodify/genomic-analyses_in_apache-spark/master/vcf_filtering_tutorial/Cbp31_SNPs_test0.01.vcf.gz [following]\n--2019-06-24 20:47:08--  https://raw.githubusercontent.com/evodify/genomic-analyses_in_apache-spark/master/vcf_filtering_tutorial/Cbp31_SNPs_test0.01.vcf.gz\nResolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.40.133\nConnecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.40.133|:443... connected.\nHTTP request sent, awaiting response... 200 OK\nLength: 27186220 (26M) [application/octet-stream]\nSaving to: ‘Cbp31_SNPs_test0.01.vcf.gz’\n\n     0K .......... .......... .......... .......... ..........  0% 1.10M 24s\n    50K .......... .......... .......... .......... ..........  0% 3.10M 16s\n   100K .......... .......... .......... .......... ..........  0% 7.19M 12s\n   150K .......... .......... .......... .......... ..........  0% 3.63M 11s\n   200K .......... .......... .......... .......... ..........  0% 5.73M 9s\n   250K .......... .......... .......... .......... ..........  1% 4.58M 9s\n   300K .......... .......... .......... .......... ..........  1% 4.22M 8s\n   350K .......... .......... .......... .......... ..........  1% 4.83M 8s\n   400K .......... .......... .......... .......... ..........  1% 6.37M 7s\n   450K .......... .......... .......... .......... ..........  1% 4.98M 7s\n   500K .......... .......... .......... .......... ..........  2% 6.07M 7s\n   550K .......... .......... .......... .......... ..........  2% 6.72M 7s\n   600K .......... .......... .......... .......... ..........  2% 5.43M 6s\n   650K .......... .......... .......... .......... ..........  2% 5.84M 6s\n   700K .......... .......... .......... .......... ..........  2% 4.43M 6s\n   750K .......... .......... .......... .......... ..........  3% 6.61M 6s\n   800K .......... .......... .......... .......... ..........  3% 6.10M 6s\n   850K .......... .......... .......... .......... ..........  3% 7.22M 6s\n   900K .......... .......... .......... .......... ..........  3% 6.85M 6s\n   950K .......... .......... .......... .......... ..........  3% 5.71M 6s\n  1000K .......... .......... .......... .......... ..........  3% 7.60M 6s\n  1050K .......... .......... .......... .......... ..........  4% 8.33M 5s\n  1100K .......... .......... .......... .......... ..........  4% 4.85M 5s\n  1150K .......... .......... .......... .......... ..........  4% 8.37M 5s\n  1200K .......... .......... .......... .......... ..........  4% 6.51M 5s\n  1250K .......... .......... .......... .......... ..........  4% 7.73M 5s\n  1300K .......... .......... .......... .......... ..........  5% 9.41M 5s\n  1350K .......... .......... .......... .......... ..........  5% 5.93M 5s\n  1400K .......... .......... .......... .......... ..........  5% 8.09M 5s\n  1450K .......... .......... .......... .......... ..........  5% 9.43M 5s\n  1500K .......... .......... .......... .......... ..........  5% 5.85M 5s\n  1550K .......... .......... .......... .......... ..........  6% 7.91M 5s\n  1600K .......... .......... .......... .......... ..........  6% 10.5M 5s\n  1650K .......... .......... .......... .......... ..........  6% 8.72M 5s\n  1700K .......... .......... .......... .......... ..........  6% 7.43M 5s\n  1750K .......... .......... .......... .......... ..........  6% 8.59M 4s\n  1800K .......... .......... .......... .......... ..........  6% 10.8M 4s\n  1850K .......... .......... .......... .......... ..........  7% 7.99M 4s\n  1900K .......... .......... .......... .......... ..........  7% 6.98M 4s\n  1950K .......... .......... .......... .......... ..........  7% 9.24M 4s\n  2000K .......... .......... .......... .......... ..........  7% 11.5M 4s\n  2050K .......... .......... .......... .......... ..........  7% 7.28M 4s\n  2100K .......... .......... .......... .......... ..........  8% 9.01M 4s\n  2150K .......... .......... .......... .......... ..........  8% 12.8M 4s\n  2200K .......... .......... .......... .......... ..........  8% 7.62M 4s\n  2250K .......... .......... .......... .......... ..........  8% 11.1M 4s\n  2300K .......... .......... .......... .......... ..........  8% 7.64M 4s\n  2350K .......... .......... .......... .......... ..........  9% 13.9M 4s\n  2400K .......... .......... .......... .......... ..........  9% 8.36M 4s\n  2450K .......... .......... .......... .......... ..........  9% 11.9M 4s\n  2500K .......... .......... .......... .......... ..........  9% 9.21M 4s\n  2550K .......... .......... .......... .......... ..........  9% 14.8M 4s\n  2600K .......... .......... .......... .......... ..........  9% 7.47M 4s\n  2650K .......... .......... .......... .......... .......... 10% 16.4M 4s\n  2700K .......... .......... .......... .......... .......... 10% 8.97M 4s\n  2750K .......... .......... .......... .......... .......... 10% 10.6M 4s\n  2800K .......... .......... .......... .......... .......... 10% 12.7M 4s\n  2850K .......... .......... .......... .......... .......... 10% 9.62M 4s\n  2900K .......... .......... .......... .......... .......... 11% 12.9M 4s\n  2950K .......... .......... .......... .......... .......... 11% 12.5M 3s\n  3000K .......... .......... .......... .......... .......... 11% 12.2M 3s\n  3050K .......... .......... .......... .......... .......... 11% 16.2M 3s\n  3100K .......... .......... .......... .......... .......... 11% 5.99M 3s\n  3150K .......... .......... .......... .......... .......... 12% 19.0M 3s\n  3200K .......... .......... .......... .......... .......... 12% 15.8M 3s\n  3250K .......... .......... .......... .......... .......... 12% 11.2M 3s\n  3300K .......... .......... .......... .......... .......... 12% 9.18M 3s\n  3350K .......... .......... .......... .......... .......... 12% 11.2M 3s\n  3400K .......... .......... .......... .......... .......... 12% 17.9M 3s\n  3450K .......... .......... .......... .......... .......... 13% 14.6M 3s\n  3500K .......... .......... .......... .......... .......... 13% 11.1M 3s\n  3550K .......... .......... .......... .......... .......... 13% 9.38M 3s\n  3600K .......... .......... .......... .......... .......... 13% 12.5M 3s\n  3650K .......... .......... .......... .......... .......... 13% 17.9M 3s\n  3700K .......... .......... .......... .......... .......... 14% 15.6M 3s\n  3750K .......... .......... .......... .......... .......... 14% 14.3M 3s\n  3800K .......... .......... .......... .......... .......... 14% 9.97M 3s\n  3850K .......... .......... .......... .......... .......... 14% 15.9M 3s\n  3900K .......... .......... .......... .......... .......... 14% 11.2M 3s\n  3950K .......... .......... .......... .......... .......... 15% 17.8M 3s\n  4000K .......... .......... .......... .......... .......... 15% 15.2M 3s\n  4050K .......... .......... .......... .......... .......... 15% 13.3M 3s\n  4100K .......... .......... .......... .......... .......... 15% 10.6M 3s\n  4150K .......... .......... .......... .......... .......... 15% 17.7M 3s\n  4200K .......... .......... .......... .......... .......... 16% 16.9M 3s\n  4250K .......... .......... .......... .......... .......... 16% 16.4M 3s\n  4300K .......... .......... .......... .......... .......... 16% 11.6M 3s\n  4350K .......... .......... .......... .......... .......... 16% 10.2M 3s\n  4400K .......... .......... .......... .......... .......... 16% 15.9M 3s\n  4450K .......... .......... .......... .......... .......... 16% 16.1M 3s\n  4500K .......... .......... .......... .......... .......... 17% 22.6M 3s\n  4550K .......... .......... .......... .......... .......... 17% 17.8M 3s\n  4600K .......... .......... .......... .......... .......... 17% 17.6M 3s\n  4650K .......... .......... .......... .......... .......... 17% 14.3M 3s\n  4700K .......... .......... .......... .......... .......... 17% 8.80M 3s\n  4750K .......... .......... .......... .......... .......... 18% 21.2M 3s\n  4800K .......... .......... .......... .......... .......... 18% 17.0M 3s\n  4850K .......... .......... .......... .......... .......... 18% 19.4M 3s\n  4900K .......... .......... .......... .......... .......... 18% 14.6M 3s\n  4950K .......... .......... .......... .......... .......... 18% 24.4M 3s\n  5000K .......... .......... .......... .......... .......... 19% 9.93M 3s\n  5050K .......... .......... .......... .......... .......... 19% 19.8M 2s\n  5100K .......... .......... .......... .......... .......... 19% 11.2M 2s\n  5150K .......... .......... .......... .......... .......... 19% 20.4M 2s\n  5200K .......... .......... .......... .......... .......... 19% 25.6M 2s\n  5250K .......... .......... .......... .......... .......... 19% 20.2M 2s\n  5300K .......... .......... .......... .......... .......... 20% 10.1M 2s\n  5350K .......... .......... .......... .......... .......... 20% 18.3M 2s\n  5400K .......... .......... .......... .......... .......... 20% 24.9M 2s\n  5450K .......... .......... .......... .......... .......... 20% 15.4M 2s\n  5500K .......... .......... .......... .......... .......... 20% 17.0M 2s\n  5550K .......... .......... .......... .......... .......... 21% 19.1M 2s\n  5600K .......... .......... .......... .......... .......... 21% 19.5M 2s\n  5650K .......... .......... .......... .......... .......... 21% 12.2M 2s\n  5700K .......... .......... .......... .......... .......... 21% 24.8M 2s\n  5750K .......... .......... .......... .......... .......... 21% 16.9M 2s\n  5800K .......... .......... .......... .......... .......... 22% 17.0M 2s\n  5850K .......... .......... .......... .......... .......... 22% 30.5M 2s\n  5900K .......... .......... .......... .......... .......... 22% 15.4M 2s\n  5950K .......... .......... .......... .......... .......... 22% 15.5M 2s\n  6000K .......... .......... .......... .......... .......... 22% 12.7M 2s\n  6050K .......... .......... .......... .......... .......... 22% 29.1M 2s\n  6100K .......... .......... .......... .......... .......... 23% 20.4M 2s\n  6150K .......... .......... .......... .......... .......... 23% 23.0M 2s\n  6200K .......... .......... .......... .......... .......... 23% 21.0M 2s\n  6250K .......... .......... .......... .......... .......... 23% 20.2M 2s\n  6300K .......... .......... .......... .......... .......... 23% 15.5M 2s\n  6350K .......... .......... .......... .......... .......... 24% 21.5M 2s\n  6400K .......... .......... .......... .......... .......... 24% 14.3M 2s\n  6450K .......... .......... .......... .......... .......... 24% 18.0M 2s\n  6500K .......... .......... .......... .......... .......... 24% 21.1M 2s\n  6550K .......... .......... .......... .......... .......... 24% 24.3M 2s\n  6600K .......... .......... .......... .......... .......... 25% 20.7M 2s\n  6650K .......... .......... .......... .......... .......... 25% 28.5M 2s\n  6700K .......... .......... .......... .......... .......... 25% 17.3M 2s\n  6750K .......... .......... .......... .......... .......... 25% 20.2M 2s\n  6800K .......... .......... .......... .......... .......... 25% 15.3M 2s\n  6850K .......... .......... .......... .......... .......... 25% 28.3M 2s\n  6900K .......... .......... .......... .......... .......... 26% 14.3M 2s\n  6950K .......... .......... .......... .......... .......... 26% 38.3M 2s\n  7000K .......... .......... .......... .......... .......... 26% 15.8M 2s\n  7050K .......... .......... .......... .......... .......... 26% 30.1M 2s\n  7100K .......... .......... .......... .......... .......... 26% 19.2M 2s\n  7150K .......... .......... .......... .......... .......... 27% 21.2M 2s\n  7200K .......... .......... .......... .......... .......... 27% 30.0M 2s\n  7250K .......... .......... .......... .......... .......... 27% 13.0M 2s\n  7300K .......... .......... .......... .......... .......... 27% 21.6M 2s\n  7350K .......... .......... .......... .......... .......... 27% 25.5M 2s\n  7400K .......... .......... .......... .......... .......... 28% 21.4M 2s\n  7450K .......... .......... .......... .......... .......... 28% 23.1M 2s\n  7500K .......... .......... .......... .......... .......... 28% 23.2M 2s\n  7550K .......... .......... .......... .......... .......... 28% 24.4M 2s\n  7600K .......... .......... .......... .......... .......... 28% 30.6M 2s\n  7650K .......... .......... .......... .......... .......... 29% 27.7M 2s\n  7700K .......... .......... .......... .......... .......... 29% 12.5M 2s\n  7750K .......... .......... .......... .......... .......... 29% 24.1M 2s\n  7800K .......... .......... .......... .......... .......... 29% 24.6M 2s\n  7850K .......... .......... .......... .......... .......... 29% 21.9M 2s\n  7900K .......... .......... .......... .......... .......... 29% 23.0M 2s\n  7950K .......... .......... .......... .......... .......... 30% 23.7M 2s\n  8000K .......... .......... .......... .......... .......... 30% 26.9M 2s\n  8050K .......... .......... .......... .......... .......... 30% 26.7M 2s\n  8100K .......... .......... .......... .......... .......... 30% 40.0M 2s\n  8150K .......... .......... .......... .......... .......... 30% 15.1M 2s\n  8200K .......... .......... .......... .......... .......... 31% 24.3M 2s\n  8250K .......... .......... .......... .......... .......... 31% 18.5M 2s\n  8300K .......... .......... .......... .......... .......... 31% 16.8M 2s\n  8350K .......... .......... .......... .......... .......... 31% 61.1M 2s\n  8400K .......... .......... .......... .......... .......... 31% 23.4M 2s\n  8450K .......... .......... .......... .......... .......... 32% 29.8M 2s\n  8500K .......... .......... .......... .......... .......... 32% 23.3M 2s\n  8550K .......... .......... .......... .......... .......... 32% 37.6M 2s\n  8600K .......... .......... .......... .......... .......... 32% 27.1M 2s\n  8650K .......... .......... .......... .......... .......... 32% 20.2M 2s\n  8700K .......... .......... .......... .......... .......... 32% 13.7M 2s\n  8750K .......... .......... .......... .......... .......... 33% 28.1M 2s\n  8800K .......... .......... .......... .......... .......... 33% 19.6M 2s\n  8850K .......... .......... .......... .......... .......... 33% 42.4M 2s\n  8900K .......... .......... .......... .......... .......... 33% 23.3M 2s\n  8950K .......... .......... .......... .......... .......... 33% 38.4M 2s\n  9000K .......... .......... .......... .......... .......... 34% 25.3M 2s\n  9050K .......... .......... .......... .......... .......... 34% 33.5M 2s\n  9100K .......... .......... .......... .......... .......... 34% 25.7M 1s\n  9150K .......... .......... .......... .......... .......... 34% 13.8M 1s\n  9200K .......... .......... .......... .......... .......... 34% 27.0M 1s\n  9250K .......... .......... .......... .......... .......... 35% 24.1M 1s\n  9300K .......... .......... .......... .......... .......... 35% 22.4M 1s\n  9350K .......... .......... .......... .......... .......... 35% 30.6M 1s\n  9400K .......... .......... .......... .......... .......... 35% 49.1M 1s\n  9450K .......... .......... .......... .......... .......... 35% 23.4M 1s\n  9500K .......... .......... .......... .......... .......... 35% 26.9M 1s\n  9550K .......... .......... .......... .......... .......... 36% 41.9M 1s\n  9600K .......... .......... .......... .......... .......... 36% 26.0M 1s\n  9650K .......... .......... .......... .......... .......... 36% 41.1M 1s\n  9700K .......... .......... .......... .......... .......... 36% 26.6M 1s\n  9750K .......... .......... .......... .......... .......... 36% 16.3M 1s\n  9800K .......... .......... .......... .......... .......... 37% 22.0M 1s\n  9850K .......... .......... .......... .......... .......... 37% 25.3M 1s\n  9900K .......... .......... .......... .......... .......... 37% 29.0M 1s\n  9950K .......... .......... .......... .......... .......... 37% 23.9M 1s\n 10000K .......... .......... .......... .......... .......... 37% 70.0M 1s\n 10050K .......... .......... .......... .......... .......... 38% 23.5M 1s\n 10100K .......... .......... .......... .......... .......... 38% 41.3M 1s\n 10150K .......... .......... .......... .......... .......... 38% 31.3M 1s\n 10200K .......... .......... .......... .......... .......... 38% 42.0M 1s\n 10250K .......... .......... .......... .......... .......... 38% 27.8M 1s\n 10300K .......... .......... .......... .......... .......... 38% 14.2M 1s\n 10350K .......... .......... .......... .......... .......... 39% 35.1M 1s\n 10400K .......... .......... .......... .......... .......... 39% 17.7M 1s\n 10450K .......... .......... .......... .......... .......... 39% 63.2M 1s\n 10500K .......... .......... .......... .......... .......... 39% 28.3M 1s\n 10550K .......... .......... .......... .......... .......... 39% 29.2M 1s\n 10600K .......... .......... .......... .......... .......... 40% 43.1M 1s\n 10650K .......... .......... .......... .......... .......... 40% 35.8M 1s\n 10700K .......... .......... .......... .......... .......... 40% 24.8M 1s\n 10750K .......... .......... .......... .......... .......... 40% 31.9M 1s\n 10800K .......... .......... .......... .......... .......... 40% 55.8M 1s\n 10850K .......... .......... .......... .......... .......... 41% 27.8M 1s\n 10900K .......... .......... .......... .......... .......... 41% 37.2M 1s\n 10950K .......... .......... .......... .......... .......... 41% 24.7M 1s\n 11000K .......... .......... .......... .......... .......... 41% 22.1M 1s\n 11050K .......... .......... .......... .......... .......... 41% 24.4M 1s\n 11100K .......... .......... .......... .......... .......... 41% 26.5M 1s\n 11150K .......... .......... .......... .......... .......... 42% 26.7M 1s\n 11200K .......... .......... .......... .......... .......... 42% 37.1M 1s\n 11250K .......... .......... .......... .......... .......... 42% 26.6M 1s\n 11300K .......... .......... .......... .......... .......... 42% 44.8M 1s\n 11350K .......... .......... .......... .......... .......... 42% 49.3M 1s\n 11400K .......... .......... .......... .......... .......... 43% 29.8M 1s\n 11450K .......... .......... .......... .......... .......... 43% 35.8M 1s\n 11500K .......... .......... .......... .......... .......... 43% 29.3M 1s\n 11550K .......... .......... .......... .......... .......... 43% 35.0M 1s\n 11600K .......... .......... .......... .......... .......... 43% 19.1M 1s\n 11650K .......... .......... .......... .......... .......... 44% 26.4M 1s\n 11700K .......... .......... .......... .......... .......... 44% 44.8M 1s\n 11750K .......... .......... .......... .......... .......... 44% 39.9M 1s\n 11800K .......... .......... .......... .......... .......... 44% 28.0M 1s\n 11850K .......... .......... .......... .......... .......... 44% 36.8M 1s\n 11900K .......... .......... .......... .......... .......... 45% 25.2M 1s\n 11950K .......... .......... .......... .......... .......... 45% 41.3M 1s\n 12000K .......... .......... .......... .......... .......... 45% 65.8M 1s\n 12050K .......... .......... .......... .......... .......... 45% 25.7M 1s\n 12100K .......... .......... .......... .......... .......... 45% 38.7M 1s\n 12150K .......... .......... .......... .......... .......... 45% 43.9M 1s\n 12200K .......... .......... .......... .......... .......... 46% 55.9M 1s\n 12250K .......... .......... .......... .......... .......... 46% 13.0M 1s\n 12300K .......... .......... .......... .......... .......... 46% 30.6M 1s\n 12350K .......... .......... .......... .......... .......... 46% 45.0M 1s\n 12400K .......... .......... .......... .......... .......... 46% 33.8M 1s\n 12450K .......... .......... .......... .......... .......... 47% 36.7M 1s\n 12500K .......... .......... .......... .......... .......... 47% 39.3M 1s\n 12550K .......... .......... .......... .......... .......... 47% 56.6M 1s\n 12600K .......... .......... .......... .......... .......... 47% 21.3M 1s\n 12650K .......... .......... .......... .......... .......... 47% 81.4M 1s\n 12700K .......... .......... .......... .......... .......... 48% 25.1M 1s\n 12750K .......... .......... .......... .......... .......... 48% 36.9M 1s\n 12800K .......... .......... .......... .......... .......... 48% 58.2M 1s\n 12850K .......... .......... .......... .......... .......... 48% 38.5M 1s\n 12900K .......... .......... .......... .......... .......... 48% 14.7M 1s\n 12950K .......... .......... .......... .......... .......... 48% 69.3M 1s\n 13000K .......... .......... .......... .......... .......... 49% 28.5M 1s\n 13050K .......... .......... .......... .......... .......... 49% 48.3M 1s\n 13100K .......... .......... .......... .......... .......... 49% 36.5M 1s\n 13150K .......... .......... .......... .......... .......... 49% 29.2M 1s\n 13200K .......... .......... .......... .......... .......... 49% 57.0M 1s\n 13250K .......... .......... .......... .......... .......... 50% 55.3M 1s\n 13300K .......... .......... .......... .......... .......... 50% 22.3M 1s\n 13350K .......... .......... .......... .......... .......... 50% 45.8M 1s\n 13400K .......... .......... .......... .......... .......... 50% 36.2M 1s\n 13450K .......... .......... .......... .......... .......... 50% 59.0M 1s\n 13500K .......... .......... .......... .......... .......... 51% 32.4M 1s\n 13550K .......... .......... .......... .......... .......... 51% 36.8M 1s\n 13600K .......... .......... .......... .......... .......... 51% 77.9M 1s\n 13650K .......... .......... .......... .......... .......... 51% 15.5M 1s\n 13700K .......... .......... .......... .......... .......... 51% 24.3M 1s\n 13750K .......... .......... .......... .......... .......... 51% 71.8M 1s\n 13800K .......... .......... .......... .......... .......... 52% 37.9M 1s\n 13850K .......... .......... .......... .......... .......... 52% 41.9M 1s\n 13900K .......... .......... .......... .......... .......... 52% 35.3M 1s\n 13950K .......... .......... .......... .......... .......... 52% 70.0M 1s\n 14000K .......... .......... .......... .......... .......... 52% 32.2M 1s\n 14050K .......... .......... .......... .......... .......... 53% 43.2M 1s\n 14100K .......... .......... .......... .......... .......... 53% 40.7M 1s\n 14150K .......... .......... .......... .......... .......... 53% 30.8M 1s\n 14200K .......... .......... .......... .......... .......... 53% 79.8M 1s\n 14250K .......... .......... .......... .......... .......... 53% 37.7M 1s\n 14300K .......... .......... .......... .......... .......... 54% 46.2M 1s\n 14350K .......... .......... .......... .......... .......... 54% 88.3M 1s\n 14400K .......... .......... .......... .......... .......... 54% 13.3M 1s\n 14450K .......... .......... .......... .......... .......... 54% 27.5M 1s\n 14500K .......... .......... .......... .......... .......... 54% 55.8M 1s\n 14550K .......... .......... .......... .......... .......... 54% 58.4M 1s\n 14600K .......... .......... .......... .......... .......... 55% 33.6M 1s\n 14650K .......... .......... .......... .......... .......... 55% 70.8M 1s\n 14700K .......... .......... .......... .......... .......... 55% 39.0M 1s\n 14750K .......... .......... .......... .......... .......... 55% 46.7M 1s\n 14800K .......... .......... .......... .......... .......... 55% 35.7M 1s\n 14850K .......... .......... .......... .......... .......... 56% 33.8M 1s\n 14900K .......... .......... .......... .......... .......... 56% 33.2M 1s\n 14950K .......... .......... .......... .......... .......... 56% 51.3M 1s\n 15000K .......... .......... .......... .......... .......... 56% 36.0M 1s\n 15050K .......... .......... .......... .......... .......... 56% 54.6M 1s\n 15100K .......... .......... .......... .......... .......... 57% 80.8M 1s\n 15150K .......... .......... .......... .......... .......... 57% 38.3M 1s\n 15200K .......... .......... .......... .......... .......... 57% 20.2M 1s\n 15250K .......... .......... .......... .......... .......... 57% 29.4M 1s\n 15300K .......... .......... .......... .......... .......... 57% 60.7M 1s\n 15350K .......... .......... .......... .......... .......... 58% 35.7M 1s\n 15400K .......... .......... .......... .......... .......... 58% 39.0M 1s\n 15450K .......... .......... .......... .......... .......... 58% 40.0M 1s\n 15500K .......... .......... .......... .......... .......... 58% 97.7M 1s\n 15550K .......... .......... .......... .......... .......... 58% 56.3M 1s\n 15600K .......... .......... .......... .......... .......... 58% 34.7M 1s\n 15650K .......... .......... .......... .......... .......... 59% 33.8M 1s\n 15700K .......... .......... .......... .......... .......... 59% 90.5M 1s\n 15750K .......... .......... .......... .......... .......... 59% 28.4M 1s\n 15800K .......... .......... .......... .......... .......... 59% 74.6M 1s\n 15850K .......... .......... .......... .......... .......... 59% 35.6M 1s\n 15900K .......... .......... .......... .......... .......... 60% 28.1M 1s\n 15950K .......... .......... .......... .......... .......... 60% 84.6M 1s\n 16000K .......... .......... .......... .......... .......... 60%  119M 1s\n 16050K .......... .......... .......... .......... .......... 60% 22.1M 1s\n 16100K .......... .......... .......... .......... .......... 60% 20.1M 1s\n 16150K .......... .......... .......... .......... .......... 61% 44.5M 1s\n 16200K .......... .......... .......... .......... .......... 61% 88.3M 1s\n 16250K .......... .......... .......... .......... .......... 61% 39.0M 1s\n 16300K .......... .......... .......... .......... .......... 61% 34.3M 1s\n 16350K .......... .......... .......... .......... .......... 61% 72.8M 1s\n 16400K .......... .......... .......... .......... .......... 61% 43.1M 1s\n 16450K .......... .......... .......... .......... .......... 62% 61.3M 1s\n 16500K .......... .......... .......... .......... .......... 62% 31.5M 1s\n 16550K .......... .......... .......... .......... .......... 62% 89.4M 1s\n 16600K .......... .......... .......... .......... .......... 62% 35.5M 1s\n 16650K .......... .......... .......... .......... .......... 62% 38.8M 1s\n 16700K .......... .......... .......... .......... .......... 63% 71.9M 1s\n 16750K .......... .......... .......... .......... .......... 63% 39.1M 1s\n 16800K .......... .......... .......... .......... .......... 63% 47.5M 1s\n 16850K .......... .......... .......... .......... .......... 63% 71.2M 1s\n 16900K .......... .......... .......... .......... .......... 63% 22.1M 1s\n 16950K .......... .......... .......... .......... .......... 64%  110M 1s\n 17000K .......... .......... .......... .......... .......... 64% 23.7M 1s\n 17050K .......... .......... .......... .......... .......... 64% 41.1M 1s\n 17100K .......... .......... .......... .......... .......... 64% 33.4M 1s\n 17150K .......... .......... .......... .......... .......... 64% 94.6M 1s\n 17200K .......... .......... .......... .......... .......... 64% 42.4M 1s\n 17250K .......... .......... .......... .......... .......... 65% 82.0M 1s\n 17300K .......... .......... .......... .......... .......... 65% 48.5M 1s\n 17350K .......... .......... .......... .......... .......... 65% 24.3M 1s\n 17400K .......... .......... .......... .......... .......... 65% 94.4M 1s\n 17450K .......... .......... .......... .......... .......... 65%  103M 1s\n 17500K .......... .......... .......... .......... .......... 66% 38.8M 1s\n 17550K .......... .......... .......... .......... .......... 66% 34.4M 1s\n 17600K .......... .......... .......... .......... .......... 66% 33.9M 1s\n 17650K .......... .......... .......... .......... .......... 66% 91.5M 1s\n 17700K .......... .......... .......... .......... .......... 66% 66.0M 1s\n 17750K .......... .......... .......... .......... .......... 67% 56.5M 1s\n 17800K .......... .......... .......... .......... .......... 67% 23.7M 1s\n 17850K .......... .......... .......... .......... .......... 67%  105M 0s\n 17900K .......... .......... .......... .......... .......... 67% 22.8M 0s\n 17950K .......... .......... .......... .......... .......... 67% 43.8M 0s\n 18000K .......... .......... .......... .......... .......... 67% 82.9M 0s\n 18050K .......... .......... .......... .......... .......... 68% 40.4M 0s\n 18100K .......... .......... .......... .......... .......... 68% 66.9M 0s\n 18150K .......... .......... .......... .......... .......... 68% 64.0M 0s\n 18200K .......... .......... .......... .......... .......... 68% 75.1M 0s\n 18250K .......... .......... .......... .......... .......... 68% 47.8M 0s\n 18300K .......... .......... .......... .......... .......... 69% 51.5M 0s\n 18350K .......... .......... .......... .......... .......... 69% 45.9M 0s\n 18400K .......... .......... .......... .......... .......... 69% 81.2M 0s\n 18450K .......... .......... .......... .......... .......... 69% 29.4M 0s\n 18500K .......... .......... .......... .......... .......... 69% 41.8M 0s\n 18550K .......... .......... .......... .......... .......... 70%  120M 0s\n 18600K .......... .......... .......... .......... .......... 70% 36.1M 0s\n 18650K .......... .......... .......... .......... .......... 70% 52.4M 0s\n 18700K .......... .......... .......... .......... .......... 70% 69.4M 0s\n 18750K .......... .......... .......... .......... .......... 70% 21.8M 0s\n 18800K .......... .......... .......... .......... .......... 71%  102M 0s\n 18850K .......... .......... .......... .......... .......... 71% 26.6M 0s\n 18900K .......... .......... .......... .......... .......... 71% 34.0M 0s\n 18950K .......... .......... .......... .......... .......... 71% 83.0M 0s\n 19000K .......... .......... .......... .......... .......... 71% 40.2M 0s\n 19050K .......... .......... .......... .......... .......... 71% 85.5M 0s\n 19100K .......... .......... .......... .......... .......... 72% 53.6M 0s\n 19150K .......... .......... .......... .......... .......... 72% 77.8M 0s\n 19200K .......... .......... .......... .......... .......... 72% 54.5M 0s\n 19250K .......... .......... .......... .......... .......... 72% 69.6M 0s\n 19300K .......... .......... .......... .......... .......... 72% 33.4M 0s\n 19350K .......... .......... .......... .......... .......... 73% 67.7M 0s\n 19400K .......... .......... .......... .......... .......... 73% 56.2M 0s\n 19450K .......... .......... .......... .......... .......... 73% 48.3M 0s\n 19500K .......... .......... .......... .......... .......... 73% 90.7M 0s\n 19550K .......... .......... .......... .......... .......... 73% 65.6M 0s\n 19600K .......... .......... .......... .......... .......... 74% 38.0M 0s\n 19650K .......... .......... .......... .......... .......... 74% 49.0M 0s\n 19700K .......... .......... .......... .......... .......... 74%  101M 0s\n 19750K .......... .......... .......... .......... .......... 74% 44.2M 0s\n 19800K .......... .......... .......... .......... .......... 74% 33.3M 0s\n 19850K .......... .......... .......... .......... .......... 74% 24.0M 0s\n 19900K .......... .......... .......... .......... .......... 75% 35.1M 0s\n 19950K .......... .......... .......... .......... .......... 75% 74.7M 0s\n 20000K .......... .......... .......... .......... .......... 75%  130M 0s\n 20050K .......... .......... .......... .......... .......... 75% 43.4M 0s\n 20100K .......... .......... .......... .......... .......... 75% 48.1M 0s\n 20150K .......... .......... .......... .......... .......... 76% 69.7M 0s\n 20200K .......... .......... .......... .......... .......... 76%  117M 0s\n 20250K .......... .......... .......... .......... .......... 76% 47.1M 0s\n 20300K .......... .......... .......... .......... .......... 76% 28.7M 0s\n 20350K .......... .......... .......... .......... .......... 76% 73.5M 0s\n 20400K .......... .......... .......... .......... .......... 77% 61.6M 0s\n 20450K .......... .......... .......... .......... .......... 77%  106M 0s\n 20500K .......... .......... .......... .......... .......... 77% 63.6M 0s\n 20550K .......... .......... .......... .......... .......... 77% 40.2M 0s\n 20600K .......... .......... .......... .......... .......... 77% 41.8M 0s\n 20650K .......... .......... .......... .......... .......... 77% 80.5M 0s\n 20700K .......... .......... .......... .......... .......... 78% 92.3M 0s\n 20750K .......... .......... .......... .......... .......... 78% 54.2M 0s\n 20800K .......... .......... .......... .......... .......... 78% 52.1M 0s\n 20850K .......... .......... .......... .......... .......... 78% 32.0M 0s\n 20900K .......... .......... .......... .......... .......... 78%  108M 0s\n 20950K .......... .......... .......... .......... .......... 79% 32.9M 0s\n 21000K .......... .......... .......... .......... .......... 79% 37.7M 0s\n 21050K .......... .......... .......... .......... .......... 79% 81.8M 0s\n 21100K .......... .......... .......... .......... .......... 79% 37.3M 0s\n 21150K .......... .......... .......... .......... .......... 79%  162M 0s\n 21200K .......... .......... .......... .......... .......... 80% 51.6M 0s\n 21250K .......... .......... .......... .......... .......... 80% 55.7M 0s\n 21300K .......... .......... .......... .......... .......... 80% 50.2M 0s\n 21350K .......... .......... .......... .......... .......... 80%  164M 0s\n 21400K .......... .......... .......... .......... .......... 80% 26.9M 0s\n 21450K .......... .......... .......... .......... .......... 80%  109M 0s\n 21500K .......... .......... .......... .......... .......... 81% 45.6M 0s\n 21550K .......... .......... .......... .......... .......... 81% 63.0M 0s\n 21600K .......... .......... .......... .......... .......... 81%  164M 0s\n 21650K .......... .......... .......... .......... .......... 81% 39.7M 0s\n 21700K .......... .......... .......... .......... .......... 81% 43.2M 0s\n 21750K .......... .......... .......... .......... .......... 82% 68.0M 0s\n 21800K .......... .......... .......... .......... .......... 82% 78.8M 0s\n 21850K .......... .......... .......... .......... .......... 82% 70.4M 0s\n 21900K .......... .......... .......... .......... .......... 82% 46.7M 0s\n 21950K .......... .......... .......... .......... .......... 82% 36.5M 0s\n 22000K .......... .......... .......... .......... .......... 83% 28.0M 0s\n 22050K .......... .......... .......... .......... .......... 83%  178M 0s\n 22100K .......... .......... .......... .......... .......... 83% 37.6M 0s\n 22150K .......... .......... .......... .......... .......... 83% 95.0M 0s\n 22200K .......... .......... .......... .......... .......... 83% 61.9M 0s\n 22250K .......... .......... .......... .......... .......... 83% 48.5M 0s\n 22300K .......... .......... .......... .......... .......... 84%  172M 0s\n 22350K .......... .......... .......... .......... .......... 84% 63.6M 0s\n 22400K .......... .......... .......... .......... .......... 84% 54.1M 0s\n 22450K .......... .......... .......... .......... .......... 84% 56.1M 0s\n 22500K .......... .......... .......... .......... .......... 84% 29.7M 0s\n 22550K .......... .......... .......... .......... .......... 85%  134M 0s\n 22600K .......... .......... .......... .......... .......... 85% 91.1M 0s\n 22650K .......... .......... .......... .......... .......... 85% 51.1M 0s\n 22700K .......... .......... .......... .......... .......... 85% 62.0M 0s\n 22750K .......... .......... .......... .......... .......... 85% 77.0M 0s\n 22800K .......... .......... .......... .......... .......... 86% 56.6M 0s\n 22850K .......... .......... .......... .......... .......... 86% 42.5M 0s\n 22900K .......... .......... .......... .......... .......... 86% 62.3M 0s\n 22950K .......... .......... .......... .......... .......... 86%  132M 0s\n 23000K .......... .......... .......... .......... .......... 86% 58.8M 0s\n 23050K .......... .......... .......... .......... .......... 87% 39.0M 0s\n 23100K .......... .......... .......... .......... .......... 87% 39.8M 0s\n 23150K .......... .......... .......... .......... .......... 87% 29.9M 0s\n 23200K .......... .......... .......... .......... .......... 87% 42.8M 0s\n 23250K .......... .......... .......... .......... .......... 87%  145M 0s\n 23300K .......... .......... .......... .......... .......... 87% 87.6M 0s\n 23350K .......... .......... .......... .......... .......... 88% 59.6M 0s\n 23400K .......... .......... .......... .......... .......... 88% 57.5M 0s\n 23450K .......... .......... .......... .......... .......... 88%  199M 0s\n 23500K .......... .......... .......... .......... .......... 88% 48.3M 0s\n 23550K .......... .......... .......... .......... .......... 88% 88.5M 0s\n 23600K .......... .......... .......... .......... .......... 89% 54.2M 0s\n 23650K .......... .......... .......... .......... .......... 89% 60.0M 0s\n 23700K .......... .......... .......... .......... .......... 89%  142M 0s\n 23750K .......... .......... .......... .......... .......... 89% 44.4M 0s\n 23800K .......... .......... .......... .......... .......... 89% 93.1M 0s\n 23850K .......... .......... .......... .......... .......... 90% 46.6M 0s\n 23900K .......... .......... .......... .......... .......... 90% 35.7M 0s\n 23950K .......... .......... .......... .......... .......... 90% 90.7M 0s\n 24000K .......... .......... .......... .......... .......... 90%  147M 0s\n 24050K .......... .......... .......... .......... .......... 90% 42.9M 0s\n 24100K .......... .......... .......... .......... .......... 90% 77.4M 0s\n 24150K .......... .......... .......... .......... .......... 91%  155M 0s\n 24200K .......... .......... .......... .......... .......... 91% 82.1M 0s\n 24250K .......... .......... .......... .......... .......... 91% 37.9M 0s\n 24300K .......... .......... .......... .......... .......... 91% 29.7M 0s\n 24350K .......... .......... .......... .......... .......... 91% 31.7M 0s\n 24400K .......... .......... .......... .......... .......... 92%  167M 0s\n 24450K .......... .......... .......... .......... .......... 92% 38.1M 0s\n 24500K .......... .......... .......... .......... .......... 92% 93.8M 0s\n 24550K .......... .......... .......... .......... .......... 92% 56.9M 0s\n 24600K .......... .......... .......... .......... .......... 92% 82.9M 0s\n 24650K .......... .......... .......... .......... .......... 93%  129M 0s\n 24700K .......... .......... .......... .......... .......... 93% 50.9M 0s\n 24750K .......... .......... .......... .......... .......... 93% 73.7M 0s\n 24800K .......... .......... .......... .......... .......... 93% 62.3M 0s\n 24850K .......... .......... .......... .......... .......... 93% 70.6M 0s\n 24900K .......... .......... .......... .......... .......... 93%  146M 0s\n 24950K .......... .......... .......... .......... .......... 94% 61.2M 0s\n 25000K .......... .......... .......... .......... .......... 94% 58.8M 0s\n 25050K .......... .......... .......... .......... .......... 94% 67.5M 0s\n 25100K .......... .......... .......... .......... .......... 94% 69.6M 0s\n 25150K .......... .......... .......... .......... .......... 94%  136M 0s\n 25200K .......... .......... .......... .......... .......... 95% 47.2M 0s\n 25250K .......... .......... .......... .......... .......... 95% 69.5M 0s\n 25300K .......... .......... .......... .......... .......... 95% 44.3M 0s\n 25350K .......... .......... .......... .......... .......... 95% 66.0M 0s\n 25400K .......... .......... .......... .......... .......... 95%  131M 0s\n 25450K .......... .......... .......... .......... .......... 96% 93.6M 0s\n 25500K .......... .......... .......... .......... .......... 96% 85.7M 0s\n 25550K .......... .......... .......... .......... .......... 96% 20.8M 0s\n 25600K .......... .......... .......... .......... .......... 96%  191M 0s\n 25650K .......... .......... .......... .......... .......... 96% 31.9M 0s\n 25700K .......... .......... .......... .......... .......... 96% 37.5M 0s\n 25750K .......... .......... .......... .......... .......... 97% 89.1M 0s\n 25800K .......... .......... .......... .......... .......... 97% 62.8M 0s\n 25850K .......... .......... .......... .......... .......... 97%  110M 0s\n 25900K .......... .......... .......... .......... .......... 97% 79.3M 0s\n 25950K .......... .......... .......... .......... .......... 97% 52.9M 0s\n 26000K .......... .......... .......... .......... .......... 98%  105M 0s\n 26050K .......... .......... .......... .......... .......... 98%  159M 0s\n 26100K .......... .......... .......... .......... .......... 98% 58.3M 0s\n 26150K .......... .......... .......... .......... .......... 98% 64.3M 0s\n 26200K .......... .......... .......... .......... .......... 98%  107M 0s\n 26250K .......... .......... .......... .......... .......... 99% 65.7M 0s\n 26300K .......... .......... .......... .......... .......... 99% 73.5M 0s\n 26350K .......... .......... .......... .......... .......... 99%  132M 0s\n 26400K .......... .......... .......... .......... .......... 99% 67.0M 0s\n 26450K .......... .......... .......... .......... .......... 99% 71.3M 0s\n 26500K .......... .......... .......... .......... ......... 100%  164M=1.2s\n\n2019-06-24 20:47:10 (21.9 MB/s) - ‘Cbp31_SNPs_test0.01.vcf.gz’ saved [27186220/27186220]\n\n--2019-06-24 20:47:10--  https://github.com/evodify/genomic-analyses_in_apache-spark/raw/master/vcf_filtering_tutorial/Cbp31_annot.csv\nResolving github.com (github.com)... 192.30.255.112\nConnecting to github.com (github.com)|192.30.255.112|:443... connected.\nHTTP request sent, awaiting response... 302 Found\nLocation: https://raw.githubusercontent.com/evodify/genomic-analyses_in_apache-spark/master/vcf_filtering_tutorial/Cbp31_annot.csv [following]\n--2019-06-24 20:47:10--  https://raw.githubusercontent.com/evodify/genomic-analyses_in_apache-spark/master/vcf_filtering_tutorial/Cbp31_annot.csv\nResolving raw.githubusercontent.com (raw.githubusercontent.com)... 151.101.40.133\nConnecting to raw.githubusercontent.com (raw.githubusercontent.com)|151.101.40.133|:443... connected.\nHTTP request sent, awaiting response... 200 OK\nLength: 712 [text/plain]\nSaving to: ‘Cbp31_annot.csv.2’\n\n     0K                                                       100%  111M=0s\n\n2019-06-24 20:47:11 (111 MB/s) - ‘Cbp31_annot.csv.2’ saved [712/712]\n\n</div>"]}}],"execution_count":16},{"cell_type":"code","source":["%sh pwd && ls -l"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">/databricks/driver\ntotal 104316\n-rw-r--r-- 1 root root       712 Jun 24 20:26 Cbp31_annot.csv\n-rw-r--r-- 1 root root       712 Jun 24 20:38 Cbp31_annot.csv.1\n-rw-r--r-- 1 root root       712 Jun 24 20:47 Cbp31_annot.csv.2\n-rw-r--r-- 1 root root 106677117 Jun 24 20:47 Cbp31_SNPs_test0.01.vcf\ndrwxr-xr-x 2 root root      4096 Jan  1  1970 conf\n-rw-r--r-- 1 root root       731 Jun 24 20:24 derby.log\ndrwxr-xr-x 3 root root      4096 Jun 24 20:24 eventlogs\ndrwxr-xr-x 2 root root      4096 Jun 24 20:45 ganglia\n-rw-r--r-- 1 root root    103281 Jun 24 20:47 hail.log\ndrwxr-xr-x 2 root root      4096 Jun 24 20:24 logs\n</div>"]}}],"execution_count":17},{"cell_type":"markdown","source":["## Move the downloaded data to DBFS"],"metadata":{}},{"cell_type":"code","source":["\ndbutils.fs.mkdirs(\"dbfs:/FileStore/tables/Cbp\") # create a new directory for our data files\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">59</span><span class=\"ansired\">]: </span>True\n</div>"]}}],"execution_count":19},{"cell_type":"code","source":["dir(dbutils)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">60</span><span class=\"ansired\">]: </span>\n[&apos;CredentialsHandler&apos;,\n &apos;FSHandler&apos;,\n &apos;LibraryHandler&apos;,\n &apos;NotebookHandler&apos;,\n &apos;PreviewHandler&apos;,\n &apos;SecretsHandler&apos;,\n &apos;__call__&apos;,\n &apos;__class__&apos;,\n &apos;__delattr__&apos;,\n &apos;__dict__&apos;,\n &apos;__dir__&apos;,\n &apos;__doc__&apos;,\n &apos;__eq__&apos;,\n &apos;__format__&apos;,\n &apos;__ge__&apos;,\n &apos;__getattr__&apos;,\n &apos;__getattribute__&apos;,\n &apos;__getstate__&apos;,\n &apos;__gt__&apos;,\n &apos;__hash__&apos;,\n &apos;__init__&apos;,\n &apos;__le__&apos;,\n &apos;__lt__&apos;,\n &apos;__module__&apos;,\n &apos;__ne__&apos;,\n &apos;__new__&apos;,\n &apos;__reduce__&apos;,\n &apos;__reduce_ex__&apos;,\n &apos;__repr__&apos;,\n &apos;__setattr__&apos;,\n &apos;__sizeof__&apos;,\n &apos;__str__&apos;,\n &apos;__subclasshook__&apos;,\n &apos;__weakref__&apos;,\n &apos;entry_point&apos;,\n &apos;fs&apos;,\n &apos;help&apos;,\n &apos;shell&apos;,\n &apos;widgets&apos;]\n</div>"]}}],"execution_count":20},{"cell_type":"code","source":["\ndbutils.fs.mv(\"file:/databricks/driver/Cbp31_SNPs_test0.01.vcf\", \"dbfs:/FileStore/tables/Cbp\") # move\ndbutils.fs.cp(\"file:/databricks/driver/Cbp31_annot.csv\", \"dbfs:/FileStore/tables/Cbp\") # copy because we will need non-dbfs file for R later\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">61</span><span class=\"ansired\">]: </span>True\n</div>"]}}],"execution_count":21},{"cell_type":"code","source":["\ndbutils.fs.ls(\"dbfs:/FileStore/tables/Cbp/\")\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">62</span><span class=\"ansired\">]: </span>\n[FileInfo(path=&apos;dbfs:/FileStore/tables/Cbp/Cbp31_SNPs_test0.01.vcf&apos;, name=&apos;Cbp31_SNPs_test0.01.vcf&apos;, size=106677117),\n FileInfo(path=&apos;dbfs:/FileStore/tables/Cbp/Cbp31_annot.csv&apos;, name=&apos;Cbp31_annot.csv&apos;, size=712)]\n</div>"]}}],"execution_count":22},{"cell_type":"markdown","source":["## Start processing"],"metadata":{}},{"cell_type":"code","source":["vcf_path = '/FileStore/tables/Cbp/Cbp31_SNPs_test0.01.vcf'\nannotation_path = '/FileStore/tables/Cbp/Cbp31_annot.csv'\n\n# comment out the line above and uncommend the lines below for human data\n\n# vcf_path = '/databricks-datasets/hail/data-001/1kg_sample.vcf.bgz'\n# annotation_path = '/databricks-datasets/hail/data-001/1kg_annotations.txt'\n# purcell_5k_path = '/databricks-datasets/hail/data-001/purcell5k.interval_list'"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":24},{"cell_type":"code","source":["vds = hc.import_vcf(vcf_path) # bgz import fails even with force_bgz=True. I compressed my files with bgzip from https://github.com/samtools/tabix ."],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":25},{"cell_type":"markdown","source":["This method produced a [VariantDataset](https://hail.is/hail/hail.VariantDataset.html), Hail's primary representation of genomic data. Following that link to Hail's python API documentation will let you see the myriad methods it offers.  We will use but a few of them in this tutorial. We next use its [split_multi](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.split_multi) to split multi-allelic variants into biallelic variants. For example, the variant `1:1000:A:T,C` would become two variants: `1:1000:A:T` and `1:1000:A:C`."],"metadata":{}},{"cell_type":"code","source":["vds = vds.split_multi()"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":27},{"cell_type":"code","source":["print(type(vds))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">&lt;class &apos;hail.dataset.VariantDataset&apos;&gt;\n</div>"]}}],"execution_count":28},{"cell_type":"code","source":["\ndir(vds)\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">67</span><span class=\"ansired\">]: </span>\n[&apos;__class__&apos;,\n &apos;__delattr__&apos;,\n &apos;__dict__&apos;,\n &apos;__dir__&apos;,\n &apos;__doc__&apos;,\n &apos;__eq__&apos;,\n &apos;__format__&apos;,\n &apos;__ge__&apos;,\n &apos;__getattribute__&apos;,\n &apos;__gt__&apos;,\n &apos;__hash__&apos;,\n &apos;__init__&apos;,\n &apos;__le__&apos;,\n &apos;__lt__&apos;,\n &apos;__module__&apos;,\n &apos;__ne__&apos;,\n &apos;__new__&apos;,\n &apos;__reduce__&apos;,\n &apos;__reduce_ex__&apos;,\n &apos;__repr__&apos;,\n &apos;__setattr__&apos;,\n &apos;__sizeof__&apos;,\n &apos;__str__&apos;,\n &apos;__subclasshook__&apos;,\n &apos;__weakref__&apos;,\n &apos;_global_schema&apos;,\n &apos;_globals&apos;,\n &apos;_jvds&apos;,\n &apos;_num_samples&apos;,\n &apos;_sa_schema&apos;,\n &apos;_sample_annotations&apos;,\n &apos;_sample_ids&apos;,\n &apos;_typecheck&apos;,\n &apos;_va_schema&apos;,\n &apos;aggregate_by_key&apos;,\n &apos;aggregate_intervals&apos;,\n &apos;annotate_alleles_expr&apos;,\n &apos;annotate_global_list&apos;,\n &apos;annotate_global_py&apos;,\n &apos;annotate_global_table&apos;,\n &apos;annotate_samples_expr&apos;,\n &apos;annotate_samples_fam&apos;,\n &apos;annotate_samples_list&apos;,\n &apos;annotate_samples_table&apos;,\n &apos;annotate_samples_vds&apos;,\n &apos;annotate_variants_bed&apos;,\n &apos;annotate_variants_expr&apos;,\n &apos;annotate_variants_intervals&apos;,\n &apos;annotate_variants_keytable&apos;,\n &apos;annotate_variants_loci&apos;,\n &apos;annotate_variants_table&apos;,\n &apos;annotate_variants_vds&apos;,\n &apos;cache&apos;,\n &apos;concordance&apos;,\n &apos;count&apos;,\n &apos;count_variants&apos;,\n &apos;deduplicate&apos;,\n &apos;downsample_variants&apos;,\n &apos;export_gen&apos;,\n &apos;export_genotypes&apos;,\n &apos;export_plink&apos;,\n &apos;export_samples&apos;,\n &apos;export_variants&apos;,\n &apos;export_variants_cass&apos;,\n &apos;export_variants_solr&apos;,\n &apos;export_vcf&apos;,\n &apos;file_version&apos;,\n &apos;filter_alleles&apos;,\n &apos;filter_genotypes&apos;,\n &apos;filter_multi&apos;,\n &apos;filter_samples_all&apos;,\n &apos;filter_samples_expr&apos;,\n &apos;filter_samples_list&apos;,\n &apos;filter_variants_all&apos;,\n &apos;filter_variants_expr&apos;,\n &apos;filter_variants_intervals&apos;,\n &apos;filter_variants_list&apos;,\n &apos;global_schema&apos;,\n &apos;globals&apos;,\n &apos;grm&apos;,\n &apos;hardcalls&apos;,\n &apos;hc&apos;,\n &apos;ibd&apos;,\n &apos;impute_sex&apos;,\n &apos;is_dosage&apos;,\n &apos;join&apos;,\n &apos;linreg&apos;,\n &apos;lmmreg&apos;,\n &apos;logreg&apos;,\n &apos;make_keytable&apos;,\n &apos;mendel_errors&apos;,\n &apos;min_rep&apos;,\n &apos;num_partitions&apos;,\n &apos;num_samples&apos;,\n &apos;pca&apos;,\n &apos;persist&apos;,\n &apos;query_samples&apos;,\n &apos;query_samples_typed&apos;,\n &apos;query_variants&apos;,\n &apos;query_variants_typed&apos;,\n &apos;rename_samples&apos;,\n &apos;repartition&apos;,\n &apos;same&apos;,\n &apos;sample_annotations&apos;,\n &apos;sample_ids&apos;,\n &apos;sample_qc&apos;,\n &apos;sample_schema&apos;,\n &apos;samples_keytable&apos;,\n &apos;sparkinfo&apos;,\n &apos;split_multi&apos;,\n &apos;tdt&apos;,\n &apos;variant_qc&apos;,\n &apos;variant_schema&apos;,\n &apos;variants_keytable&apos;,\n &apos;vep&apos;,\n &apos;was_split&apos;,\n &apos;write&apos;]\n</div>"]}}],"execution_count":29},{"cell_type":"markdown","source":["We next use the [annotate_samples_table](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.annotate_samples_table) method to load phenotypic information on each sample from the sample annotations file.\n\nHere `annotation_path` refers to the sample annotation data file, whose first few lines are:\n```\nSample Population\nDL174     ASI\nGY37      ASI\nHJC419    ASI\n12.4      EUR\n13.16     EUR\n16.9      EUR\nBEL5      EUR\n```\n\nThe `root` argument says where to put this data. For sample annotations, the root must start with `sa` followed by a `.` and the rest is up to you, so let's use `sa.myAnnot`.\n\nThe `sample_expr` argument indicates that the sample ID is in column `Sample`.\n\nThe object `TextTableConfig` allows users to provide information about column data types, header existence, comment characters, and field delimiters. For example,  'Population: Boolean'.  `impute=True` will infer column types automatically."],"metadata":{}},{"cell_type":"code","source":["dbutils.fs.head(annotation_path)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">68</span><span class=\"ansired\">]: </span>&apos;Sample\\tPopulation\\tLatitude\\tLongitude\\nDL174\\tASI\\t38.56\\t121.35\\nGY37\\tASI\\t26.37\\t106.43\\nHJC419\\tASI\\t30.16\\t120.13\\nHY85\\tASI\\t26.53\\t112.33\\nJZH152\\tASI\\t30.20\\t112.06\\nKMB206\\tASI\\t25.06\\t102.41\\nNJ219\\tASI\\t32.03\\t118.46\\nPL\\tASI\\t23.97\\t120.95\\nTBS195\\tASI\\t33.57\\t107.45\\nTY118\\tASI\\t37.55\\t112.32\\nXN444\\tASI\\t36.37\\t101.46\\n12.4\\tEUR\\t51.50\\t11.97\\n13.16\\tEUR\\t50.06\\t19.94\\n16.9\\tEUR\\t51.81\\t5.84\\nBEL5\\tEUR\\t50.55\\t128.28\\nFR50\\tEUR\\t48.08\\t7.37\\nHRB132\\tEUR\\t45.45\\t126.37\\nIRRU2\\tEUR\\t52.16\\t104.18\\nRK32\\tEUR\\t64.13\\t-21.82\\nSE14\\tEUR\\t62.64\\t17.94\\nSE33\\tEUR\\t56.15\\t13.77\\nSTA4\\tEUR\\t56.20\\t-2.47\\nSTJ2\\tEUR\\t44.51\\t-1.21\\nVLA3\\tEUR\\t43.13\\t131.40\\n39-12-28\\tME\\t44.40\\t7.40\\n5.16\\tME\\t41.65\\t-4.72\\n70.5\\tME\\t37.97\\t24.00\\nAL87\\tME\\t35.45\\t7.96\\nJO56\\tME\\t31.97\\t35.98\\nTR73\\tME\\t41.02\\t28.97\\nWAC5\\tME\\t31.29\\t-97.17\\n&apos;\n</div>"]}}],"execution_count":31},{"cell_type":"code","source":["\nvds = vds.annotate_samples_table(annotation_path,\n                                 root='sa.myAnnot',\n                                 sample_expr='Sample',\n                                 config=TextTableConfig(impute=True))\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":32},{"cell_type":"markdown","source":["Lastly, we'll [write](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.write) the dataset to disk so that all future computations begin by reading in the fast VDS rather than the slow VCF."],"metadata":{}},{"cell_type":"code","source":["out_path = '/cbp.vds'\nvds.write(out_path, overwrite=True)\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":34},{"cell_type":"markdown","source":["## Start exploring\n\nNow we're ready to start exploring! We will read back in the VDS we wrote to disk:"],"metadata":{}},{"cell_type":"code","source":["vds"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">71</span><span class=\"ansired\">]: </span>&lt;hail.dataset.VariantDataset at 0x7f468c109c18&gt;\n</div>"]}}],"execution_count":36},{"cell_type":"code","source":["hc.sc"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">72</span><span class=\"ansired\">]: </span>&lt;SparkContext master=local[8] appName=Databricks Shell&gt;\n</div>"]}}],"execution_count":37},{"cell_type":"code","source":["print(out_path)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">/cbp.vds\n</div>"]}}],"execution_count":38},{"cell_type":"code","source":["import py4j"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":39},{"cell_type":"code","source":["vds = hc.read(out_path)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">FatalError</span>                                Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398378&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">----&gt; 1</span><span class=\"ansiyellow\"> </span>vds <span class=\"ansiyellow\">=</span> hc<span class=\"ansiyellow\">.</span>read<span class=\"ansiyellow\">(</span>out_path<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/context.py</span> in <span class=\"ansicyan\">read</span><span class=\"ansiblue\">(self, path, sites_only)</span>\n<span class=\"ansigreen\">    454</span>         <span class=\"ansigreen\">if</span> sites_only<span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    455</span>             pargs<span class=\"ansiyellow\">.</span>append<span class=\"ansiyellow\">(</span><span class=\"ansiblue\">&quot;--skip-genotypes&quot;</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">--&gt; 456</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">return</span> self<span class=\"ansiyellow\">.</span>_run_command<span class=\"ansiyellow\">(</span><span class=\"ansigreen\">None</span><span class=\"ansiyellow\">,</span> pargs<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    457</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    458</span>     <span class=\"ansigreen\">def</span> write_partitioning<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">,</span> path<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/context.py</span> in <span class=\"ansicyan\">_run_command</span><span class=\"ansiblue\">(self, vds, pargs)</span>\n<span class=\"ansigreen\">     77</span>             result <span class=\"ansiyellow\">=</span> cmd<span class=\"ansiyellow\">.</span>run<span class=\"ansiyellow\">(</span>jstate<span class=\"ansiyellow\">,</span> cmd_args<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">     78</span>         <span class=\"ansigreen\">except</span> Py4JJavaError <span class=\"ansigreen\">as</span> e<span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">---&gt; 79</span><span class=\"ansiyellow\">             </span>raise_py4j_exception<span class=\"ansiyellow\">(</span>e<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">     80</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">     81</span>         <span class=\"ansigreen\">return</span> VariantDataset<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">,</span> result<span class=\"ansiyellow\">.</span>vds<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/java.py</span> in <span class=\"ansicyan\">raise_py4j_exception</span><span class=\"ansiblue\">(e)</span>\n<span class=\"ansigreen\">     85</span> <span class=\"ansigreen\">def</span> raise_py4j_exception<span class=\"ansiyellow\">(</span>e<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">     86</span>     msg <span class=\"ansiyellow\">=</span> env<span class=\"ansiyellow\">.</span>jutils<span class=\"ansiyellow\">.</span>getMinimalMessage<span class=\"ansiyellow\">(</span>e<span class=\"ansiyellow\">.</span>java_exception<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">---&gt; 87</span><span class=\"ansiyellow\">     </span><span class=\"ansigreen\">raise</span> FatalError<span class=\"ansiyellow\">(</span>msg<span class=\"ansiyellow\">,</span> e<span class=\"ansiyellow\">.</span>java_exception<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">FatalError</span>: \ncorrupt VDS: invalid metadata file.\n  Recreate VDS with current version of Hail.\n  caught exception: java.lang.NoSuchMethodError: org.json4s.jackson.JsonMethods$.parse(Lorg/json4s/JsonInput;Z)Lorg/json4s/JsonAST$JValue;\n\tat is.hail.variant.VariantSampleMatrix$$anonfun$1.apply(VariantSampleMatrix.scala:65)\n\tat is.hail.variant.VariantSampleMatrix$$anonfun$1.apply(VariantSampleMatrix.scala:65)\n\tat is.hail.utils.richUtils.RichHadoopConfiguration$.readFile$extension(RichHadoopConfiguration.scala:243)\n\tat is.hail.variant.VariantSampleMatrix$.readMetadata(VariantSampleMatrix.scala:65)\n\tat is.hail.variant.VariantSampleMatrix$.read(VariantSampleMatrix.scala:171)\n\tat is.hail.driver.Read$$anonfun$1.apply(Read.scala:41)\n\tat is.hail.driver.Read$$anonfun$1.apply(Read.scala:41)\n\tat scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)\n\tat scala.collection.TraversableLike$$anonfun$map$1.apply(TraversableLike.scala:234)\n\tat scala.collection.IndexedSeqOptimized$class.foreach(IndexedSeqOptimized.scala:33)\n\tat scala.collection.mutable.ArrayOps$ofRef.foreach(ArrayOps.scala:186)\n\tat scala.collection.TraversableLike$class.map(TraversableLike.scala:234)\n\tat scala.collection.mutable.ArrayOps$ofRef.map(ArrayOps.scala:186)\n\tat is.hail.driver.Read$.run(Read.scala:41)\n\tat is.hail.driver.Read$.run(Read.scala:9)\n\tat is.hail.driver.Command.runCommand(Command.scala:256)\n\tat is.hail.driver.Command.run(Command.scala:261)\n\tat sun.reflect.GeneratedMethodAccessor434.invoke(Unknown Source)\n\tat sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)\n\tat java.lang.reflect.Method.invoke(Method.java:498)\n\tat py4j.reflection.MethodInvoker.invoke(MethodInvoker.java:244)\n\tat py4j.reflection.ReflectionEngine.invoke(ReflectionEngine.java:380)\n\tat py4j.Gateway.invoke(Gateway.java:295)\n\tat py4j.commands.AbstractCommand.invokeMethod(AbstractCommand.java:132)\n\tat py4j.commands.CallCommand.execute(CallCommand.java:79)\n\tat py4j.GatewayConnection.run(GatewayConnection.java:251)\n\tat java.lang.Thread.run(Thread.java:748)\n         </div>"]}}],"execution_count":40},{"cell_type":"code","source":[""],"metadata":{},"outputs":[],"execution_count":41},{"cell_type":"markdown","source":["First, we'll print some statistics about the size of the dataset using [count](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.count):"],"metadata":{}},{"cell_type":"code","source":["print(vds.count())"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">{&apos;nSamples&apos;: 31, &apos;nVariants&apos;: 98094, &apos;nGenotypes&apos;: 3040914}\n</div>"]}}],"execution_count":43},{"cell_type":"markdown","source":["If the Boolean parameter `genotypes` is set to `True`, the overall call rate across all genotypes is computed as well:"],"metadata":{}},{"cell_type":"code","source":["vds.count(genotypes=True)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">81</span><span class=\"ansired\">]: </span>{&apos;nGenotypes&apos;: 3040914, &apos;nVariants&apos;: 98094, &apos;nSamples&apos;: 31, &apos;nCalled&apos;: 2574591, &apos;callRate&apos;: 84.66503820890692}\n</div>"]}}],"execution_count":45},{"cell_type":"markdown","source":["So the call rate before any QC filtering is about 84.67%.\n\nLet's print the types of all annotations.\n\nVariant annotations:"],"metadata":{}},{"cell_type":"code","source":["print(vds.variant_schema)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Struct {\n    rsid: String,\n    qual: Double,\n    filters: Set[String],\n    pass: Boolean,\n    info: Struct {\n        AC: Array[Int],\n        AF: Array[Double],\n        AN: Int,\n        BaseQRankSum: Double,\n        ClippingRankSum: Double,\n        DP: Int,\n        DS: Boolean,\n        END: Int,\n        ExcessHet: Double,\n        FS: Double,\n        HaplotypeScore: Double,\n        InbreedingCoeff: Double,\n        MLEAC: Array[Int],\n        MLEAF: Array[Double],\n        MQ: Double,\n        MQRankSum: Double,\n        QD: Double,\n        RAW_MQ: Double,\n        ReadPosRankSum: Double,\n        SOR: Double\n    },\n    aIndex: Int,\n    wasSplit: Boolean\n}\n</div>"]}}],"execution_count":47},{"cell_type":"markdown","source":["Sample annotations:"],"metadata":{}},{"cell_type":"code","source":["print(vds.sample_schema)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Struct {\n    myAnnot: Struct {\n        Sample: String,\n        Population: String,\n        Latitude: Double,\n        Longitude: Double\n    }\n}\n</div>"]}}],"execution_count":49},{"cell_type":"markdown","source":["We have just printed the sample and variant structure above. Global structure is empty.\n \n To recall the structure:\n\n<img src=\"https://hail.is/docs/stable/_images/hail-vds-rep.png\" width=\"600\">\n\nAlso, note the annotations imported from the original VCF, as well as the sample annotations added above. Notice how those six sample annotations loaded above are nested inside `sa.structure` as defined by the `root` option in [annotate_samples_table](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.annotate_samples_table)."],"metadata":{}},{"cell_type":"markdown","source":["Next we'll add some global annotations including the list of populations that are present in our dataset and the number of samples in each population, using the Hail expression language and the [query_samples](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.query_samples) method. The codings are:\n\n  - ASI = Asian\n  - EUR = European\n  - ME = Middle Eastern\n\nWe'll first build up a list of query expressions, then evaluate them all at once to save time."],"metadata":{}},{"cell_type":"code","source":["expressions = ['samples.map(s => sa.myAnnot.Population).collect().toSet']\nqueries = vds.query_samples(expressions)    # popping ! ! ! !  something wrong...\n\nprint('populations = %s' % queries[0])\nprint('total samples = %s' % vds.num_samples)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">NameError</span>                                 Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398389&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">      1</span> expressions <span class=\"ansiyellow\">=</span> <span class=\"ansiyellow\">[</span><span class=\"ansiblue\">&apos;samples.map(s =&gt; sa.myAnnot.Population).collect().toSet&apos;</span><span class=\"ansiyellow\">]</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">----&gt; 2</span><span class=\"ansiyellow\"> </span>queries <span class=\"ansiyellow\">=</span> vds<span class=\"ansiyellow\">.</span>query_samples<span class=\"ansiyellow\">(</span>expressions<span class=\"ansiyellow\">)</span>    <span class=\"ansired\"># popping ! ! ! !  something wrong...</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      3</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      4</span> print<span class=\"ansiyellow\">(</span><span class=\"ansiblue\">&apos;populations = %s&apos;</span> <span class=\"ansiyellow\">%</span> queries<span class=\"ansiyellow\">[</span><span class=\"ansicyan\">0</span><span class=\"ansiyellow\">]</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      5</span> print<span class=\"ansiyellow\">(</span><span class=\"ansiblue\">&apos;total samples = %s&apos;</span> <span class=\"ansiyellow\">%</span> vds<span class=\"ansiyellow\">.</span>num_samples<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/dataset.py</span> in <span class=\"ansicyan\">query_samples</span><span class=\"ansiblue\">(self, exprs)</span>\n<span class=\"ansigreen\">   2797</span>         <span class=\"ansiyellow\">:</span>rtype<span class=\"ansiyellow\">:</span> list<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2798</span>         &quot;&quot;&quot;\n<span class=\"ansigreen\">-&gt; 2799</span><span class=\"ansiyellow\">         </span>r<span class=\"ansiyellow\">,</span> t <span class=\"ansiyellow\">=</span> self<span class=\"ansiyellow\">.</span>query_samples_typed<span class=\"ansiyellow\">(</span>exprs<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2800</span>         <span class=\"ansigreen\">return</span> r<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2801</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/dataset.py</span> in <span class=\"ansicyan\">query_samples_typed</span><span class=\"ansiblue\">(self, exprs)</span>\n<span class=\"ansigreen\">   2761</span>         result_list <span class=\"ansiyellow\">=</span> self<span class=\"ansiyellow\">.</span>_jvds<span class=\"ansiyellow\">.</span>querySamples<span class=\"ansiyellow\">(</span>jarray<span class=\"ansiyellow\">(</span>env<span class=\"ansiyellow\">.</span>jvm<span class=\"ansiyellow\">.</span>java<span class=\"ansiyellow\">.</span>lang<span class=\"ansiyellow\">.</span>String<span class=\"ansiyellow\">,</span> exprs<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2762</span>         ptypes <span class=\"ansiyellow\">=</span> <span class=\"ansiyellow\">[</span>Type<span class=\"ansiyellow\">.</span>_from_java<span class=\"ansiyellow\">(</span>x<span class=\"ansiyellow\">.</span>_2<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span> <span class=\"ansigreen\">for</span> x <span class=\"ansigreen\">in</span> result_list<span class=\"ansiyellow\">]</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">-&gt; 2763</span><span class=\"ansiyellow\">         </span>annotations <span class=\"ansiyellow\">=</span> <span class=\"ansiyellow\">[</span>ptypes<span class=\"ansiyellow\">[</span>i<span class=\"ansiyellow\">]</span><span class=\"ansiyellow\">.</span>_convert_to_py<span class=\"ansiyellow\">(</span>result_list<span class=\"ansiyellow\">[</span>i<span class=\"ansiyellow\">]</span><span class=\"ansiyellow\">.</span>_1<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span> <span class=\"ansigreen\">for</span> i <span class=\"ansigreen\">in</span> xrange<span class=\"ansiyellow\">(</span>len<span class=\"ansiyellow\">(</span>ptypes<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">]</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2764</span>         <span class=\"ansigreen\">return</span> annotations<span class=\"ansiyellow\">,</span> ptypes<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2765</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">NameError</span>: name &apos;xrange&apos; is not defined</div>"]}}],"execution_count":52},{"cell_type":"markdown","source":["Now it's easy to count samples by population using the [counter](https://hail.is/expr_lang.html#counter) aggregator:"],"metadata":{}},{"cell_type":"code","source":["counter = vds.query_samples('samples.map(s => sa.myAnnot.Population).counter()')[0]\nfor x in counter:\n    print('population %s found %s times' % (x.key, x.count))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">NameError</span>                                 Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398391&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">----&gt; 1</span><span class=\"ansiyellow\"> </span>counter <span class=\"ansiyellow\">=</span> vds<span class=\"ansiyellow\">.</span>query_samples<span class=\"ansiyellow\">(</span><span class=\"ansiblue\">&apos;samples.map(s =&gt; sa.myAnnot.Population).counter()&apos;</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">[</span><span class=\"ansicyan\">0</span><span class=\"ansiyellow\">]</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      2</span> <span class=\"ansigreen\">for</span> x <span class=\"ansigreen\">in</span> counter<span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      3</span>     print<span class=\"ansiyellow\">(</span><span class=\"ansiblue\">&apos;population %s found %s times&apos;</span> <span class=\"ansiyellow\">%</span> <span class=\"ansiyellow\">(</span>x<span class=\"ansiyellow\">.</span>key<span class=\"ansiyellow\">,</span> x<span class=\"ansiyellow\">.</span>count<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/dataset.py</span> in <span class=\"ansicyan\">query_samples</span><span class=\"ansiblue\">(self, exprs)</span>\n<span class=\"ansigreen\">   2797</span>         <span class=\"ansiyellow\">:</span>rtype<span class=\"ansiyellow\">:</span> list<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2798</span>         &quot;&quot;&quot;\n<span class=\"ansigreen\">-&gt; 2799</span><span class=\"ansiyellow\">         </span>r<span class=\"ansiyellow\">,</span> t <span class=\"ansiyellow\">=</span> self<span class=\"ansiyellow\">.</span>query_samples_typed<span class=\"ansiyellow\">(</span>exprs<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2800</span>         <span class=\"ansigreen\">return</span> r<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2801</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/dataset.py</span> in <span class=\"ansicyan\">query_samples_typed</span><span class=\"ansiblue\">(self, exprs)</span>\n<span class=\"ansigreen\">   2761</span>         result_list <span class=\"ansiyellow\">=</span> self<span class=\"ansiyellow\">.</span>_jvds<span class=\"ansiyellow\">.</span>querySamples<span class=\"ansiyellow\">(</span>jarray<span class=\"ansiyellow\">(</span>env<span class=\"ansiyellow\">.</span>jvm<span class=\"ansiyellow\">.</span>java<span class=\"ansiyellow\">.</span>lang<span class=\"ansiyellow\">.</span>String<span class=\"ansiyellow\">,</span> exprs<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2762</span>         ptypes <span class=\"ansiyellow\">=</span> <span class=\"ansiyellow\">[</span>Type<span class=\"ansiyellow\">.</span>_from_java<span class=\"ansiyellow\">(</span>x<span class=\"ansiyellow\">.</span>_2<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span> <span class=\"ansigreen\">for</span> x <span class=\"ansigreen\">in</span> result_list<span class=\"ansiyellow\">]</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">-&gt; 2763</span><span class=\"ansiyellow\">         </span>annotations <span class=\"ansiyellow\">=</span> <span class=\"ansiyellow\">[</span>ptypes<span class=\"ansiyellow\">[</span>i<span class=\"ansiyellow\">]</span><span class=\"ansiyellow\">.</span>_convert_to_py<span class=\"ansiyellow\">(</span>result_list<span class=\"ansiyellow\">[</span>i<span class=\"ansiyellow\">]</span><span class=\"ansiyellow\">.</span>_1<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span> <span class=\"ansigreen\">for</span> i <span class=\"ansigreen\">in</span> xrange<span class=\"ansiyellow\">(</span>len<span class=\"ansiyellow\">(</span>ptypes<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">]</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2764</span>         <span class=\"ansigreen\">return</span> annotations<span class=\"ansiyellow\">,</span> ptypes<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2765</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">NameError</span>: name &apos;xrange&apos; is not defined</div>"]}}],"execution_count":54},{"cell_type":"markdown","source":["## Quality control (QC)\n\nVCF file contains many annotations scores that define the quality of genotypes as well as quality of a variant.\n\n### Filter genotypes\n\nLet's filter genotypes based on genotype quality (GQ) and read coverage (DP).\n\nHere `g` is genotype, `v` is variant, `s` is sample, and annotations are accessible via `va`, `sa`, and `global`."],"metadata":{}},{"cell_type":"code","source":["%scala\n\n// This allows easy embedding of publicly available information into any other notebook\n// when viewing in git-book just ignore this block - you may have to manually chase the URL in frameIt(\"URL\").\n// Example usage:\n// displayHTML(frameIt(\"https://en.wikipedia.org/wiki/Latent_Dirichlet_allocation#Topics_in_LDA\",250))\ndef frameIt( u:String, h:Int ) : String = {\n      \"\"\"<iframe \n src=\"\"\"\"+ u+\"\"\"\"\n width=\"95%\" height=\"\"\"\" + h + \"\"\"\"\n sandbox>\n  <p>\n    <a href=\"http://spark.apache.org/docs/latest/index.html\">\n      Fallback link for browsers that, unlikely, don't support frames\n    </a>\n  </p>\n</iframe>\"\"\"\n   }\ndisplayHTML(frameIt(\"https://en.wikipedia.org/wiki/Phred_quality_score\",500))"],"metadata":{},"outputs":[],"execution_count":56},{"cell_type":"code","source":["filter_condition_gDP_gGQ = 'g.dp >= 10 && g.gq >= 20'\nvds_gDP_gGQ = vds.filter_genotypes(filter_condition_gDP_gGQ)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":57},{"cell_type":"code","source":["vds_gDP_gGQ.count(genotypes=True)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">87</span><span class=\"ansired\">]: </span>{&apos;nGenotypes&apos;: 3040914, &apos;nVariants&apos;: 98094, &apos;nSamples&apos;: 31, &apos;nCalled&apos;: 1532341, &apos;callRate&apos;: 50.39080355445764}\n</div>"]}}],"execution_count":58},{"cell_type":"markdown","source":["Now the call rate is about 50%, so nearly 35% of genotypes failed the filter. Filtering out a genotype is equivalent to setting the genotype call to missing.\n\nHaving removed suspect genotypes, let's next remove variants with low call rate and then calculate summary statistics per sample with the [sample_qc](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.sample_qc) method."],"metadata":{}},{"cell_type":"code","source":["vds_gDP_gGQ_vCR = (vds_gDP_gGQ\n    .filter_variants_expr('gs.fraction(g => g.isCalled) >= 0.50')\n    .sample_qc())"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":60},{"cell_type":"markdown","source":["Check how many variants retained after filtering."],"metadata":{}},{"cell_type":"code","source":["vds_gDP_gGQ_vCR.count(genotypes=True)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">89</span><span class=\"ansired\">]: </span>{&apos;nGenotypes&apos;: 1549783, &apos;nVariants&apos;: 49993, &apos;nSamples&apos;: 31, &apos;nCalled&apos;: 1300202, &apos;callRate&apos;: 83.8957454043566}\n</div>"]}}],"execution_count":62},{"cell_type":"markdown","source":["### Filter samples"],"metadata":{}},{"cell_type":"markdown","source":["The call rate for each variant is calculated using the `fraction` [aggregable](https://hail.is/expr_lang.html#aggregables) on the genotypes `gs`. [sample_qc](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.sample_qc) adds a number of statistics to sample annotations:"],"metadata":{}},{"cell_type":"code","source":["print(vds_gDP_gGQ_vCR.sample_schema)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Struct {\n    myAnnot: Struct {\n        Sample: String,\n        Population: String,\n        Latitude: Double,\n        Longitude: Double\n    },\n    qc: Struct {\n        callRate: Double,\n        nCalled: Int,\n        nNotCalled: Int,\n        nHomRef: Int,\n        nHet: Int,\n        nHomVar: Int,\n        nSNP: Int,\n        nInsertion: Int,\n        nDeletion: Int,\n        nSingleton: Int,\n        nTransition: Int,\n        nTransversion: Int,\n        dpMean: Double,\n        dpStDev: Double,\n        gqMean: Double,\n        gqStDev: Double,\n        nNonRef: Int,\n        rTiTv: Double,\n        rHetHomVar: Double,\n        rInsertionDeletion: Double\n    }\n}\n</div>"]}}],"execution_count":65},{"cell_type":"markdown","source":["Let's export these sample annotations to a text file and take a look at them:"],"metadata":{}},{"cell_type":"code","source":["vds_gDP_gGQ_vCR.export_samples('file:///sampleqc.txt', 'Sample = s.id, sa.qc.*')"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":67},{"cell_type":"code","source":["%sh\nhead /sampleqc.txt | cut -f 1-8 | column -t"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Sample    callRate     nCalled  nNotCalled  nHomRef  nHet   nHomVar  nSNP\n12.4      9.08767e-01  45432    4561        16033    23357  6042     35441\n13.16     9.13688e-01  45678    4315        16738    23586  5354     34294\n16.9      9.21269e-01  46057    3936        16748    23511  5798     35107\n39-12-28  9.17848e-01  45886    4107        15673    23725  6488     36701\n5.16      9.04287e-01  45208    4785        15547    23301  6360     36021\n70.5      9.01606e-01  45074    4919        15853    23251  5970     35191\nAL87      7.41064e-01  37048    12945       13679    18281  5088     28457\nBEL5      8.56140e-01  42801    7192        15289    22488  5024     32536\nDL174     7.95871e-01  39788    10205       13824    19608  6356     32320\n</div>"]}}],"execution_count":68},{"cell_type":"markdown","source":["We can further analyze these results locally using Python's [matplotlib](http://matplotlib.org/) library. Below is an example plot of three variables (call rate, mean depth and mean GQ), along with the code that generate the plot."],"metadata":{}},{"cell_type":"code","source":["sampleqc_table = vds_gDP_gGQ_vCR.samples_keytable().to_pandas()\n\nplt.clf() # clears the entire current figure with all its axes, but leaves the window opened, such that it may be reused for other plots.\nplt.figure(figsize=(7,4)) # figure size in inches; change according to your screen size and resolution\n\nplt.subplot(1, 3, 1)\nplt.hist(sampleqc_table[\"sa.qc.callRate\"], bins=np.arange(0.4, 1.01, .05))\nplt.xlabel(\"Call Rate\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0.4, 1)\nplt.axvline(.50, color='r')\n\nplt.subplot(1, 3, 2)\nplt.hist(sampleqc_table[\"sa.qc.dpMean\"], bins=np.arange(20, 80, 10))\nplt.xlabel(\"Mean depth\")\nplt.ylabel(\"Frequency\")\nplt.xlim(10, 80)\nplt.axvline(10, color='r')\n\nplt.subplot(1, 3, 3)\nplt.hist(sampleqc_table[\"sa.qc.gqMean\"], bins=np.arange(20, 100, 10))\nplt.xlabel(\"Mean Sample GQ\")\nplt.ylabel(\"Frequency\")\nplt.axvline(30, color = 'r')\nplt.xlim(0, 100)\n\nplt.tight_layout()\nplt.show()\ndisplay()"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">AttributeError</span>                            Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398407&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">----&gt; 1</span><span class=\"ansiyellow\"> </span>sampleqc_table <span class=\"ansiyellow\">=</span> vds_gDP_gGQ_vCR<span class=\"ansiyellow\">.</span>samples_keytable<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>to_pandas<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      2</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      3</span> plt<span class=\"ansiyellow\">.</span>clf<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span> <span class=\"ansired\"># clears the entire current figure with all its axes, but leaves the window opened, such that it may be reused for other plots.</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      4</span> plt<span class=\"ansiyellow\">.</span>figure<span class=\"ansiyellow\">(</span>figsize<span class=\"ansiyellow\">=</span><span class=\"ansiyellow\">(</span><span class=\"ansicyan\">7</span><span class=\"ansiyellow\">,</span><span class=\"ansicyan\">4</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span> <span class=\"ansired\"># figure size in inches; change according to your screen size and resolution</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      5</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/keytable.py</span> in <span class=\"ansicyan\">to_pandas</span><span class=\"ansiblue\">(self, expand, flatten)</span>\n<span class=\"ansigreen\">    533</span>         <span class=\"ansiyellow\">:</span><span class=\"ansigreen\">return</span><span class=\"ansiyellow\">:</span> pandas<span class=\"ansiyellow\">.</span>DataFrame<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    534</span>         &quot;&quot;&quot;\n<span class=\"ansigreen\">--&gt; 535</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">return</span> self<span class=\"ansiyellow\">.</span>to_dataframe<span class=\"ansiyellow\">(</span>expand<span class=\"ansiyellow\">,</span> flatten<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>toPandas<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    536</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    537</span>     <span class=\"ansigreen\">def</span> export_mongodb<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">,</span> mode<span class=\"ansiyellow\">=</span><span class=\"ansiblue\">&apos;append&apos;</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/databricks/spark/python/pyspark/sql/dataframe.py</span> in <span class=\"ansicyan\">toPandas</span><span class=\"ansiblue\">(self)</span>\n<span class=\"ansigreen\">   2127</span>         <span class=\"ansigreen\">import</span> pandas <span class=\"ansigreen\">as</span> pd<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2128</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">-&gt; 2129</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">if</span> self<span class=\"ansiyellow\">.</span>sql_ctx<span class=\"ansiyellow\">.</span>_conf<span class=\"ansiyellow\">.</span>pandasRespectSessionTimeZone<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2130</span>             timezone <span class=\"ansiyellow\">=</span> self<span class=\"ansiyellow\">.</span>sql_ctx<span class=\"ansiyellow\">.</span>_conf<span class=\"ansiyellow\">.</span>sessionLocalTimeZone<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2131</span>         <span class=\"ansigreen\">else</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/databricks/spark/python/pyspark/sql/context.py</span> in <span class=\"ansicyan\">_conf</span><span class=\"ansiblue\">(self)</span>\n<span class=\"ansigreen\">     97</span>     <span class=\"ansigreen\">def</span> _conf<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">     98</span>         <span class=\"ansiblue\">&quot;&quot;&quot;Accessor for the JVM SQL-specific configurations&quot;&quot;&quot;</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">---&gt; 99</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">return</span> self<span class=\"ansiyellow\">.</span>sparkSession<span class=\"ansiyellow\">.</span>_jsparkSession<span class=\"ansiyellow\">.</span>sessionState<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>conf<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    100</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    101</span>     <span class=\"ansiyellow\">@</span>classmethod<span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">AttributeError</span>: &apos;JavaMember&apos; object has no attribute &apos;sessionState&apos;</div>"]}}],"execution_count":70},{"cell_type":"markdown","source":["You can remove samples that are outliers in the plots above, where cutoffs are given by the red lines. But there are no outliers here. If we had to filter, we could do this step:"],"metadata":{}},{"cell_type":"code","source":["vds_gDP_gGQ_vCR_sDP_sGT = (vds_gDP_gGQ_vCR\n    .annotate_samples_vds(vds_gDP_gGQ_vCR, code = 'sa.qc = vds.qc' )\n    .filter_samples_expr('sa.qc.dpMean > 0.50 && sa.qc.dpMean >=10 && sa.qc.gqMean >= 20 '))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":72},{"cell_type":"markdown","source":["As before, we can count the number of samples that remain in the dataset after filtering. (But nothing has been filtered out here)"],"metadata":{}},{"cell_type":"code","source":["vds_gDP_gGQ_vCR_sDP_sGT.count(genotypes=True)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">95</span><span class=\"ansired\">]: </span>{&apos;nGenotypes&apos;: 1549783, &apos;nVariants&apos;: 49993, &apos;nSamples&apos;: 31, &apos;nCalled&apos;: 1300202, &apos;callRate&apos;: 83.8957454043566}\n</div>"]}}],"execution_count":74},{"cell_type":"markdown","source":["### Filter variants\n\nWe now have `vds_gDP_gGQ_vCR_sDP_sGT`, a VDS where low-quality genotypes and samples have been removed.\n\nLet's use the [variant_qc](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.variant_qc) method to start exploring variant metrics:"],"metadata":{}},{"cell_type":"code","source":["vds_gDP_gGQ_vCR_sDP_sGT = vds_gDP_gGQ_vCR_sDP_sGT.variant_qc()\nprint(vds_gDP_gGQ_vCR_sDP_sGT.variant_schema)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">Struct {\n    rsid: String,\n    qual: Double,\n    filters: Set[String],\n    pass: Boolean,\n    info: Struct {\n        AC: Array[Int],\n        AF: Array[Double],\n        AN: Int,\n        BaseQRankSum: Double,\n        ClippingRankSum: Double,\n        DP: Int,\n        DS: Boolean,\n        END: Int,\n        ExcessHet: Double,\n        FS: Double,\n        HaplotypeScore: Double,\n        InbreedingCoeff: Double,\n        MLEAC: Array[Int],\n        MLEAF: Array[Double],\n        MQ: Double,\n        MQRankSum: Double,\n        QD: Double,\n        RAW_MQ: Double,\n        ReadPosRankSum: Double,\n        SOR: Double\n    },\n    aIndex: Int,\n    wasSplit: Boolean,\n    qc: Struct {\n        callRate: Double,\n        AC: Int,\n        AF: Double,\n        nCalled: Int,\n        nNotCalled: Int,\n        nHomRef: Int,\n        nHet: Int,\n        nHomVar: Int,\n        dpMean: Double,\n        dpStDev: Double,\n        gqMean: Double,\n        gqStDev: Double,\n        nNonRef: Int,\n        rHeterozygosity: Double,\n        rHetHomVar: Double,\n        rExpectedHetFrequency: Double,\n        pHWE: Double\n    }\n}\n</div>"]}}],"execution_count":76},{"cell_type":"markdown","source":["Next, we will filter variants following the [Best Practices GATK recommendations](https://gatkforums.broadinstitute.org/gatk/discussion/2806/howto-apply-hard-filters-to-a-call-set).\n\nThese recommendations are for human data, but our data is not human and the distribution of quality statistics will differ from expected for human. (Explain why it is so is beyond the scope of this tutorial.)\n\nLet's have a look at the distribution of different variant quality statistics:\n\n- QD - variant confidence standardized by depth.\n\n  This annotation puts the variant confidence QUAL score into perspective by normalizing for the amount of coverage available. Because each read contributes a little to the QUAL score, variants in region with deep coverage can have artificially inflated QUAL scores, giving the impression that the call is supported by more evidence than it really is. To compensate for this, we normalize the variant confidence by depth, which gives us a more objective picture of how well supported the call is.\n   \n- MQ - Mapping quality of a SNP.\n\n- FS - strand bias in support for REF vs ALT allele calls.\n\n  Strand bias is a type of sequencing bias in which one DNA strand is favored over the other, which can result in incorrect evaluation of the amount of evidence observed for one allele vs. the other. The FisherStrand annotation is one of several methods that aims to evaluate whether there is strand bias in the data. It uses Fisher's Exact Test to determine if there is strand bias between forward and reverse strands for the reference or alternate allele. The output is a Phred-scaled p-value. The higher the output value, the more likely there is to be bias. More bias is indicative of false positive calls.\n  \n- SOR - sequencing bias in which one DNA strand is favored over the other\n   \n   Strand bias is a type of sequencing bias in which one DNA strand is favored over the other, which can result in incorrect evaluation of the amount of evidence observed for one allele vs. the other. It is used to determine if there is strand bias between forward and reverse strands for the reference or alternate allele. The reported value is ln-scaled.\n   \n- MQRankSum - Rank sum test for mapping qualities of REF vs. ALT reads.\n   \n   This variant-level annotation compares the mapping qualities of the reads supporting the reference allele with those supporting the alternate allele. The ideal result is a value close to zero, which indicates there is little to no difference. A negative value indicates that the reads supporting the alternate allele have lower mapping quality scores than those supporting the reference allele. Conversely, a positive value indicates that the reads supporting the alternate allele have higher mapping quality scores than those supporting the reference allele.\n\n- ReadPosRankSum - do all the reads support a SNP call tend to be near the end of a read.\n   \n   The ideal result is a value close to zero, which indicates there is little to no difference in where the alleles are found relative to the ends of reads. A negative value indicates that the alternate allele is found at the ends of reads more often than the reference allele. Conversely, a positive value indicates that the reference allele is found at the ends of reads more often than the alternate allele."],"metadata":{}},{"cell_type":"markdown","source":["We've once again used matplotlib to make histograms of these siz summary statistics."],"metadata":{}},{"cell_type":"code","source":["variantqc_table = vds_gDP_gGQ_vCR_sDP_sGT.variants_keytable().to_pandas()\n\nplt.clf() # clears the entire current figure with all its axes, but leaves the window opened, such that it may be reused for other plots.\nplt.figure(figsize=(7,7)) # figure size in inches; change according to your screen size and resolution\n\nplt.subplot(3, 2, 1)\nvariantgq_MQ = variantqc_table[\"va.info.MQ\"]\nplt.hist(variantgq_MQ.dropna(), bins = np.arange(0, 101, 2)) # It is important to add dropna() to skip NA values. Otherwise, the script won't work.\nplt.xlabel(\"MQ\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 100)\nplt.axvline(30, color = 'r')\n\nplt.subplot(3, 2, 2)\nvariantgq_SOR = variantqc_table[\"va.info.SOR\"]\nplt.hist(variantgq_SOR.dropna(), bins = np.arange(0, 8, 0.2))\nplt.xlabel(\"SOR\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 8)\nplt.axvline(4, color = 'r')\n\nplt.subplot(3, 2, 3)\nvariantgq_QD = variantqc_table[\"va.info.QD\"]\nplt.hist(variantgq_QD.dropna(), bins = np.arange(0, 40, 1))\nplt.xlabel(\"QD\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 40)\nplt.axvline(2, color = 'r')\n\nplt.subplot(3, 2, 4)\nvariantgq_FS = variantqc_table[\"va.info.FS\"]\nplt.hist(variantgq_FS.dropna(), bins = np.arange(0, 100, 2))\nplt.xlabel(\"FS\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 100)\nplt.axvline(60, color = 'r')\n\nplt.subplot(3, 2, 5)\nvariantgq_MQRankSum = variantqc_table[\"va.info.MQRankSum\"]\nplt.hist(variantgq_MQRankSum.dropna(), bins = np.arange(-20, 20, 1))\nplt.xlabel(\"MQRankSum\")\nplt.ylabel(\"Frequency\")\nplt.xlim(-25, 25)\nplt.axvline(-20, color = 'r')\n\nplt.subplot(3, 2, 6)\nvariantgq_ReadPosRankSum = variantqc_table[\"va.info.ReadPosRankSum\"]\nplt.hist(variantgq_ReadPosRankSum.dropna(), bins = np.arange(-20, 20, 0.5))\nplt.xlabel(\"ReadPosRankSum\")\nplt.ylabel(\"Frequency\")\nplt.xlim(-12, 12)\nplt.axvline(-10, color = 'r')\nplt.axvline(10, color = 'r')\n\nplt.tight_layout()\nplt.show()\ndisplay()"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">AttributeError</span>                            Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398416&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">----&gt; 1</span><span class=\"ansiyellow\"> </span>variantqc_table <span class=\"ansiyellow\">=</span> vds_gDP_gGQ_vCR_sDP_sGT<span class=\"ansiyellow\">.</span>variants_keytable<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>to_pandas<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      2</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      3</span> plt<span class=\"ansiyellow\">.</span>clf<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span> <span class=\"ansired\"># clears the entire current figure with all its axes, but leaves the window opened, such that it may be reused for other plots.</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      4</span> plt<span class=\"ansiyellow\">.</span>figure<span class=\"ansiyellow\">(</span>figsize<span class=\"ansiyellow\">=</span><span class=\"ansiyellow\">(</span><span class=\"ansicyan\">7</span><span class=\"ansiyellow\">,</span><span class=\"ansicyan\">7</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span> <span class=\"ansired\"># figure size in inches; change according to your screen size and resolution</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      5</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/keytable.py</span> in <span class=\"ansicyan\">to_pandas</span><span class=\"ansiblue\">(self, expand, flatten)</span>\n<span class=\"ansigreen\">    533</span>         <span class=\"ansiyellow\">:</span><span class=\"ansigreen\">return</span><span class=\"ansiyellow\">:</span> pandas<span class=\"ansiyellow\">.</span>DataFrame<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    534</span>         &quot;&quot;&quot;\n<span class=\"ansigreen\">--&gt; 535</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">return</span> self<span class=\"ansiyellow\">.</span>to_dataframe<span class=\"ansiyellow\">(</span>expand<span class=\"ansiyellow\">,</span> flatten<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>toPandas<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    536</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    537</span>     <span class=\"ansigreen\">def</span> export_mongodb<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">,</span> mode<span class=\"ansiyellow\">=</span><span class=\"ansiblue\">&apos;append&apos;</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/databricks/spark/python/pyspark/sql/dataframe.py</span> in <span class=\"ansicyan\">toPandas</span><span class=\"ansiblue\">(self)</span>\n<span class=\"ansigreen\">   2127</span>         <span class=\"ansigreen\">import</span> pandas <span class=\"ansigreen\">as</span> pd<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2128</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">-&gt; 2129</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">if</span> self<span class=\"ansiyellow\">.</span>sql_ctx<span class=\"ansiyellow\">.</span>_conf<span class=\"ansiyellow\">.</span>pandasRespectSessionTimeZone<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2130</span>             timezone <span class=\"ansiyellow\">=</span> self<span class=\"ansiyellow\">.</span>sql_ctx<span class=\"ansiyellow\">.</span>_conf<span class=\"ansiyellow\">.</span>sessionLocalTimeZone<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2131</span>         <span class=\"ansigreen\">else</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/databricks/spark/python/pyspark/sql/context.py</span> in <span class=\"ansicyan\">_conf</span><span class=\"ansiblue\">(self)</span>\n<span class=\"ansigreen\">     97</span>     <span class=\"ansigreen\">def</span> _conf<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">     98</span>         <span class=\"ansiblue\">&quot;&quot;&quot;Accessor for the JVM SQL-specific configurations&quot;&quot;&quot;</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">---&gt; 99</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">return</span> self<span class=\"ansiyellow\">.</span>sparkSession<span class=\"ansiyellow\">.</span>_jsparkSession<span class=\"ansiyellow\">.</span>sessionState<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>conf<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    100</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    101</span>     <span class=\"ansiyellow\">@</span>classmethod<span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">AttributeError</span>: &apos;JavaMember&apos; object has no attribute &apos;sessionState&apos;</div>"]}}],"execution_count":79},{"cell_type":"markdown","source":["Lastly we use the [filter_variants_expr](https://hail.is/hail/hail.VariantDataset.html#hail.VariantDataset.filter_variants_expr) method to keep only those variants that meet the cut-off requirements (red lines in the plots above)."],"metadata":{}},{"cell_type":"code","source":["vds_gDP_gGQ_vCR_sDP_sGT_vFilter = vds_gDP_gGQ_vCR_sDP_sGT.filter_variants_expr('va.info.MQ >= 30.00 && va.info.SOR <= 4.000 && va.info.QD >= 2.00 && va.info.FS <= 60.000 && va.info.MQRankSum >= -20.000 && va.info.ReadPosRankSum >= -10.000 && va.info.ReadPosRankSum <= 10.000')\nprint('variants before filtering: %d' % vds_gDP_gGQ_vCR_sDP_sGT.count_variants())\nprint('variants after filtering: %d' % vds_gDP_gGQ_vCR_sDP_sGT_vFilter.count_variants())"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">variants before filtering: 49993\nvariants after filtering: 44368\n</div>"]}}],"execution_count":81},{"cell_type":"markdown","source":["Verify the filtering results with plots:"],"metadata":{}},{"cell_type":"code","source":["# https://discuss.hail.is/t/potential-compatibility-issue-with-spark-2-2-0/417/3"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":83},{"cell_type":"code","source":["variantqc_table = vds_gDP_gGQ_vCR_sDP_sGT_vFilter.variants_keytable().to_pandas()\n\nplt.clf() # clears the entire current figure with all its axes, but leaves the window opened, such that it may be reused for other plots.\nplt.figure(figsize=(7,7)) # figure size in inches; change according to your screen size and resolution\n\nplt.subplot(3, 2, 1)\nvariantgq_MQ = variantqc_table[\"va.info.MQ\"]\nplt.hist(variantgq_MQ.dropna(), bins = np.arange(0, 101, 2))\nplt.xlabel(\"MQ\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 100)\nplt.axvline(30, color = 'r')\n\nplt.subplot(3, 2, 2)\nvariantgq_SOR = variantqc_table[\"va.info.SOR\"]\nplt.hist(variantgq_SOR.dropna(), bins = np.arange(0, 8, 0.2))\nplt.xlabel(\"SOR\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 8)\nplt.axvline(4, color = 'r')\n\nplt.subplot(3, 2, 3)\nvariantgq_QD = variantqc_table[\"va.info.QD\"]\nplt.hist(variantgq_QD.dropna(), bins = np.arange(0, 40, 1))\nplt.xlabel(\"QD\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 40)\nplt.axvline(2, color = 'r')\n\nplt.subplot(3, 2, 4)\nvariantgq_FS = variantqc_table[\"va.info.FS\"]\nplt.hist(variantgq_FS.dropna(), bins = np.arange(0, 100, 2))\nplt.xlabel(\"FS\")\nplt.ylabel(\"Frequency\")\nplt.xlim(0, 100)\nplt.axvline(60, color = 'r')\n\nplt.subplot(3, 2, 5)\nvariantgq_MQRankSum = variantqc_table[\"va.info.MQRankSum\"]\nplt.hist(variantgq_MQRankSum.dropna(), bins = np.arange(-20, 20, 1))\nplt.xlabel(\"MQRankSum\")\nplt.ylabel(\"Frequency\")\nplt.xlim(-25, 25)\nplt.axvline(-20, color = 'r')\n\nplt.subplot(3, 2, 6)\nvariantgq_ReadPosRankSum = variantqc_table[\"va.info.ReadPosRankSum\"]\nplt.hist(variantgq_ReadPosRankSum.dropna(), bins = np.arange(-20, 20, 0.5))\nplt.xlabel(\"ReadPosRankSum\")\nplt.ylabel(\"Frequency\")\nplt.xlim(-12, 12)\nplt.axvline(-10, color = 'r')\nplt.axvline(10, color = 'r')\n\nplt.tight_layout()\nplt.show()\ndisplay()"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">AttributeError</span>                            Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398420&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">----&gt; 1</span><span class=\"ansiyellow\"> </span>variantqc_table <span class=\"ansiyellow\">=</span> vds_gDP_gGQ_vCR_sDP_sGT_vFilter<span class=\"ansiyellow\">.</span>variants_keytable<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>to_pandas<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      2</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      3</span> plt<span class=\"ansiyellow\">.</span>clf<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span> <span class=\"ansired\"># clears the entire current figure with all its axes, but leaves the window opened, such that it may be reused for other plots.</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      4</span> plt<span class=\"ansiyellow\">.</span>figure<span class=\"ansiyellow\">(</span>figsize<span class=\"ansiyellow\">=</span><span class=\"ansiyellow\">(</span><span class=\"ansicyan\">7</span><span class=\"ansiyellow\">,</span><span class=\"ansicyan\">7</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">)</span> <span class=\"ansired\"># figure size in inches; change according to your screen size and resolution</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      5</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/keytable.py</span> in <span class=\"ansicyan\">to_pandas</span><span class=\"ansiblue\">(self, expand, flatten)</span>\n<span class=\"ansigreen\">    533</span>         <span class=\"ansiyellow\">:</span><span class=\"ansigreen\">return</span><span class=\"ansiyellow\">:</span> pandas<span class=\"ansiyellow\">.</span>DataFrame<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    534</span>         &quot;&quot;&quot;\n<span class=\"ansigreen\">--&gt; 535</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">return</span> self<span class=\"ansiyellow\">.</span>to_dataframe<span class=\"ansiyellow\">(</span>expand<span class=\"ansiyellow\">,</span> flatten<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>toPandas<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    536</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    537</span>     <span class=\"ansigreen\">def</span> export_mongodb<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">,</span> mode<span class=\"ansiyellow\">=</span><span class=\"ansiblue\">&apos;append&apos;</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/databricks/spark/python/pyspark/sql/dataframe.py</span> in <span class=\"ansicyan\">toPandas</span><span class=\"ansiblue\">(self)</span>\n<span class=\"ansigreen\">   2127</span>         <span class=\"ansigreen\">import</span> pandas <span class=\"ansigreen\">as</span> pd<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2128</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">-&gt; 2129</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">if</span> self<span class=\"ansiyellow\">.</span>sql_ctx<span class=\"ansiyellow\">.</span>_conf<span class=\"ansiyellow\">.</span>pandasRespectSessionTimeZone<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2130</span>             timezone <span class=\"ansiyellow\">=</span> self<span class=\"ansiyellow\">.</span>sql_ctx<span class=\"ansiyellow\">.</span>_conf<span class=\"ansiyellow\">.</span>sessionLocalTimeZone<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">   2131</span>         <span class=\"ansigreen\">else</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansigreen\">/databricks/spark/python/pyspark/sql/context.py</span> in <span class=\"ansicyan\">_conf</span><span class=\"ansiblue\">(self)</span>\n<span class=\"ansigreen\">     97</span>     <span class=\"ansigreen\">def</span> _conf<span class=\"ansiyellow\">(</span>self<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">:</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">     98</span>         <span class=\"ansiblue\">&quot;&quot;&quot;Accessor for the JVM SQL-specific configurations&quot;&quot;&quot;</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">---&gt; 99</span><span class=\"ansiyellow\">         </span><span class=\"ansigreen\">return</span> self<span class=\"ansiyellow\">.</span>sparkSession<span class=\"ansiyellow\">.</span>_jsparkSession<span class=\"ansiyellow\">.</span>sessionState<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\">.</span>conf<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    100</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">    101</span>     <span class=\"ansiyellow\">@</span>classmethod<span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">AttributeError</span>: &apos;JavaMember&apos; object has no attribute &apos;sessionState&apos;</div>"]}}],"execution_count":84},{"cell_type":"markdown","source":["## PCA\n\nTo check if there is any genetic structure, we will use a principal component analysis (PCA)."],"metadata":{}},{"cell_type":"code","source":["vds_pca = (vds_gDP_gGQ_vCR_sDP_sGT_vFilter.pca(scores='sa.pca'))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":86},{"cell_type":"markdown","source":["We can then make a Python plot of the samples in PC space colored by population group:"],"metadata":{}},{"cell_type":"code","source":["pca_table = vds_pca.samples_keytable().to_pandas()\ncolors = {'ASI': 'green', 'EUR': 'red', 'ME': 'blue'}\nplt.clf()\nplt.figure(figsize=(7,7)) # figure size in inches; change according to your screen size and resolution\nplt.scatter(pca_table[\"sa.pca.PC1\"], pca_table[\"sa.pca.PC2\"], c = pca_table[\"sa.myAnnot.Population\"].map(colors), alpha = .7, s=100)\nplt.xlabel(\"PC1\")\nplt.ylabel(\"PC2\")\nplt.xlim(-0.7, 0.7)\nplt.ylim(-0.7, 0.7)\nlegend_entries = [mpatches.Patch(color= c, label=myAnnot) for myAnnot, c in colors.items()]\nplt.legend(handles=legend_entries, prop={'size': 15})\nplt.show()\ndisplay()"],"metadata":{},"outputs":[],"execution_count":88},{"cell_type":"markdown","source":["This plot reflects the known [population structure in *Capsella bursa-pastoris*](http://onlinelibrary.wiley.com/doi/10.1111/mec.13491/full).\n\nYou can see the location of these samples with some R code:"],"metadata":{}},{"cell_type":"markdown","source":["Install packages that are necessary to produce a geographic map:"],"metadata":{}},{"cell_type":"code","source":["%r \ninstall.packages(c(\"maps\", \"rworldmap\"))"],"metadata":{},"outputs":[],"execution_count":91},{"cell_type":"code","source":["%r\nlibrary(rworldmap)\nlibrary(maps)\n\ngeo <- read.table(\"/databricks/driver/Cbp31_annot.csv\", header = T) # read file not from dbfs\n\nnewmap <- getMap(resolution = \"hight\") # create a map object\n\npchS <- c(rep(15, 11), rep(19, 13), rep(17, 7)) # designate genetic clusters with different colours\ncolS <- c(rep(\"green\", 11), rep(\"red\", 13), rep(\"blue\", 7))\ngroupsL <- c('ASI', 'EUR', 'ME')\npchL <- c(15, 19, 17)\ncolL <- c(\"green\", \"red\", \"blue\")\n\npar(mar=c(3, 3, 2, 2), cex=1)\nplot(newmap, xlim = c(-40, 140), ylim = c(20, 50), asp = 1, bg='#DCDCDC', lty=3, lwd=0.3, col=\"#ffffff\")\nmap.axes()\npoints(x=geo$Longitude, y=geo$Latitude, pch=pchS, cex=1.2, col=colS)\nlegend(\"topright\", leg=groupsL, pch=pchL, col=colL, ncol=1, pt.cex=1.2, bg=\"#ffffff\")"],"metadata":{},"outputs":[],"execution_count":92},{"cell_type":"markdown","source":["## Summary\n\nData filtering:\n - Filter genotypes\n - Filter samples\n - Filter variants\n\n*Variants can be filtered before samples filtering if samples are of greater priority in a study.*\n\nSuch genetic data can be analyzed in various ways. A PCA is just one simple example."],"metadata":{}},{"cell_type":"markdown","source":["## Additional\n\nIt is recommended to go through the original [Analysis of 1000 Genomes with Spark and Hail](https://docs.databricks.com/spark/latest/training/1000-genomes.html).\n\nYou can also read [Hail Overview](https://www.hail.is/hail/overview.html), look through the [Hail objects](https://www.hail.is/hail/hail_objects.html) representing many core concepts in genetics, and check out the many Hail functions defined in the [Python API](https://hail.is/hail/api.html)."],"metadata":{}},{"cell_type":"code","source":[""],"metadata":{},"outputs":[],"execution_count":95},{"cell_type":"markdown","source":["# Shifting Gear - Another Sample"],"metadata":{}},{"cell_type":"code","source":["#"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"></div>"]}}],"execution_count":97},{"cell_type":"code","source":["from pprint import pprint\nhl.plot.output_notebook()"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">NameError</span>                                 Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398448&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">      1</span> <span class=\"ansigreen\">from</span> pprint <span class=\"ansigreen\">import</span> pprint<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">----&gt; 2</span><span class=\"ansiyellow\"> </span>hl<span class=\"ansiyellow\">.</span>plot<span class=\"ansiyellow\">.</span>output_notebook<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">NameError</span>: name &apos;hl&apos; is not defined</div>"]}}],"execution_count":98},{"cell_type":"code","source":[""],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">NameError</span>                                 Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398449&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">----&gt; 1</span><span class=\"ansiyellow\"> </span>hl<span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">NameError</span>: name &apos;hl&apos; is not defined</div>"]}}],"execution_count":99},{"cell_type":"code","source":["hl"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">108</span><span class=\"ansired\">]: </span>&lt;module &apos;hail&apos; from &apos;/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/__init__.py&apos;&gt;\n</div>"]}}],"execution_count":100},{"cell_type":"code","source":["hc"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">109</span><span class=\"ansired\">]: </span>&lt;hail.context.HailContext at 0x7f468c11aba8&gt;\n</div>"]}}],"execution_count":101},{"cell_type":"code","source":["\nfrom hail.plot import show\nfrom pprint import pprint\nhl.plot.output_notebook()\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">ImportError</span>                               Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398452&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">      1</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">----&gt; 2</span><span class=\"ansiyellow\"> </span><span class=\"ansigreen\">from</span> hail<span class=\"ansiyellow\">.</span>plot <span class=\"ansigreen\">import</span> show<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      3</span> <span class=\"ansigreen\">from</span> pprint <span class=\"ansigreen\">import</span> pprint<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      4</span> hl<span class=\"ansiyellow\">.</span>plot<span class=\"ansiyellow\">.</span>output_notebook<span class=\"ansiyellow\">(</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">ImportError</span>: No module named &apos;hail.plot&apos;</div>"]}}],"execution_count":102},{"cell_type":"code","source":["hl"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">111</span><span class=\"ansired\">]: </span>&lt;module &apos;hail&apos; from &apos;/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/__init__.py&apos;&gt;\n</div>"]}}],"execution_count":103},{"cell_type":"code","source":["print(type(hl))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">&lt;class &apos;module&apos;&gt;\n</div>"]}}],"execution_count":104},{"cell_type":"code","source":["hl.utils.get_1kg('data/')"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">AttributeError</span>                            Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398455&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">----&gt; 1</span><span class=\"ansiyellow\"> </span>hl<span class=\"ansiyellow\">.</span>utils<span class=\"ansiyellow\">.</span>get_1kg<span class=\"ansiyellow\">(</span><span class=\"ansiblue\">&apos;data/&apos;</span><span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">AttributeError</span>: module &apos;hail.utils&apos; has no attribute &apos;get_1kg&apos;</div>"]}}],"execution_count":105},{"cell_type":"markdown","source":["* https://hail.is/docs/0.2/tutorials/01-genome-wide-association-study.html"],"metadata":{}},{"cell_type":"code","source":["dir(hl.stats)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">114</span><span class=\"ansired\">]: </span>\n[&apos;BetaDist&apos;,\n &apos;TruncatedBetaDist&apos;,\n &apos;UniformDist&apos;,\n &apos;__builtins__&apos;,\n &apos;__cached__&apos;,\n &apos;__doc__&apos;,\n &apos;__file__&apos;,\n &apos;__loader__&apos;,\n &apos;__name__&apos;,\n &apos;__package__&apos;,\n &apos;__spec__&apos;,\n &apos;env&apos;]\n</div>"]}}],"execution_count":107},{"cell_type":"code","source":["hl.stats"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">115</span><span class=\"ansired\">]: </span>&lt;module &apos;hail.stats&apos; from &apos;/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/stats.py&apos;&gt;\n</div>"]}}],"execution_count":108},{"cell_type":"code","source":["print(type(hl.stats))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">&lt;class &apos;module&apos;&gt;\n</div>"]}}],"execution_count":109},{"cell_type":"code","source":["\nimport hail as hl\n\ndir(hl)\n\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">121</span><span class=\"ansired\">]: </span>\n[&apos;HailContext&apos;,\n &apos;KeyTable&apos;,\n &apos;TextTableConfig&apos;,\n &apos;Type&apos;,\n &apos;VariantDataset&apos;,\n &apos;__all__&apos;,\n &apos;__builtins__&apos;,\n &apos;__cached__&apos;,\n &apos;__doc__&apos;,\n &apos;__file__&apos;,\n &apos;__loader__&apos;,\n &apos;__name__&apos;,\n &apos;__package__&apos;,\n &apos;__path__&apos;,\n &apos;__spec__&apos;,\n &apos;context&apos;,\n &apos;dataset&apos;,\n &apos;java&apos;,\n &apos;keytable&apos;,\n &apos;representation&apos;,\n &apos;stats&apos;,\n &apos;type&apos;,\n &apos;utils&apos;]\n</div>"]}}],"execution_count":110},{"cell_type":"code","source":["hl.context"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">122</span><span class=\"ansired\">]: </span>&lt;module &apos;hail.context&apos; from &apos;/local_disk0/spark-c36517b3-41eb-4014-9f4a-da104c7728fd/userFiles-dbab7a83-dab7-42cf-9298-3d330d11309e/addedFile2248754259193548425c7eaedcd_8828_4b98_8cc6_90caba462aa4_hail_devel_py2_7_databricks_20ed0-8a939.egg/hail/context.py&apos;&gt;\n</div>"]}}],"execution_count":111},{"cell_type":"code","source":["print(type(hl.context))"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\">&lt;class &apos;module&apos;&gt;\n</div>"]}}],"execution_count":112},{"cell_type":"code","source":["dir(hl.utils)"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">Out[</span><span class=\"ansired\">124</span><span class=\"ansired\">]: </span>\n[&apos;Py4JJavaError&apos;,\n &apos;TextTableConfig&apos;,\n &apos;__builtins__&apos;,\n &apos;__cached__&apos;,\n &apos;__doc__&apos;,\n &apos;__file__&apos;,\n &apos;__loader__&apos;,\n &apos;__name__&apos;,\n &apos;__package__&apos;,\n &apos;__spec__&apos;,\n &apos;env&apos;,\n &apos;raise_py4j_exception&apos;]\n</div>"]}}],"execution_count":113},{"cell_type":"code","source":["\nimport hail as hl\nhl.init(sc)\n\n"],"metadata":{},"outputs":[{"metadata":{},"output_type":"display_data","data":{"text/html":["<style scoped>\n  .ansiout {\n    display: block;\n    unicode-bidi: embed;\n    white-space: pre-wrap;\n    word-wrap: break-word;\n    word-break: break-all;\n    font-family: \"Source Code Pro\", \"Menlo\", monospace;;\n    font-size: 13px;\n    color: #555;\n    margin-left: 4px;\n    line-height: 19px;\n  }\n</style>\n<div class=\"ansiout\"><span class=\"ansired\">---------------------------------------------------------------------------</span>\n<span class=\"ansired\">AttributeError</span>                            Traceback (most recent call last)\n<span class=\"ansigreen\">&lt;command-3160555503398464&gt;</span> in <span class=\"ansicyan\">&lt;module&gt;</span><span class=\"ansiblue\">()</span>\n<span class=\"ansigreen\">      1</span> <span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      2</span> <span class=\"ansigreen\">import</span> hail <span class=\"ansigreen\">as</span> hl<span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">----&gt; 3</span><span class=\"ansiyellow\"> </span>hl<span class=\"ansiyellow\">.</span>init<span class=\"ansiyellow\">(</span>sc<span class=\"ansiyellow\">)</span><span class=\"ansiyellow\"></span>\n<span class=\"ansigreen\">      4</span> <span class=\"ansiyellow\"></span>\n\n<span class=\"ansired\">AttributeError</span>: module &apos;hail&apos; has no attribute &apos;init&apos;</div>"]}}],"execution_count":114},{"cell_type":"code","source":[""],"metadata":{},"outputs":[],"execution_count":115}],"metadata":{"name":"gamma_hail_working_databricks","notebookId":3160555503398347},"nbformat":4,"nbformat_minor":0}
